{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&ensp;\n",
    "[Home Page](../../START_HERE.ipynb)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Challenge - Gene Expression Classification\n",
    "\n",
    "\n",
    "### Introduction\n",
    "\n",
    "This notebook walks through an end-to-end GPU machine learning workflow where cuDF is used for processing the data and cuML is used to train machine learning models on it. \n",
    "\n",
    "After completing this excercise, you will be able to use cuDF to load data from disk, combine tables, scale features, use one-hote encoding and even write your own GPU kernels to efficiently transform feature columns. Additionaly you will learn how to pass this data to cuML, and how to train ML models on it. The trained model is saved and it will be used for prediction.\n",
    "\n",
    "It is not required that the user is familiar with cuDF or cuML. Since our aim is to go from ETL to ML training, a detailed introduction is out of scope for this notebook. We recommend [Introduction to cuDF](../../CuDF/01-Intro_to_cuDF.ipynb) for additional information.\n",
    "\n",
    "### Problem Statement:\n",
    "We are trying to classify patients with acute myeloid leukemia (AML) and acute lymphoblastic leukemia (ALL) using machine learning (classification) algorithms. This dataset comes from a proof-of-concept study published in 1999 by Golub et al. It showed how new cases of cancer could be classified by gene expression monitoring (via DNA microarray) and thereby provided a general approach for identifying new cancer classes and assigning tumors to known classes. \n",
    "\n",
    "Here is the dataset link: https://www.kaggle.com/crawford/gene-expression."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Here is the list of exercises and modules to work on in the lab:\n",
    "\n",
    "- Convert the serial Pandas computations to CuDF operations.\n",
    "- Utilize CuML to accelerate the machine learning models.\n",
    "- Experiment with Dask to create a cluster and distribute the data and scale the operations.\n",
    "\n",
    "You will start writing code from <a href='#dask1'>here</a>, but make sure you execute the data processing blocks to understand the dataset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Data Processing\n",
    "\n",
    "The first step is downloading the dataset and putting it in the data directory, for using in this tutorial. Download the dataset here, and place it in (host/data) folder. Now we will import the necessary libraries."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NumPy Version: 1.20.3\n",
      "Scikit-Learn Version: 0.24.2\n"
     ]
    }
   ],
   "source": [
    "import numpy as np; print('NumPy Version:', np.__version__)\n",
    "import pandas as pd\n",
    "import sys\n",
    "import sklearn; print('Scikit-Learn Version:', sklearn.__version__)\n",
    "from sklearn import preprocessing \n",
    "from sklearn.utils import resample\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_curve, auc\n",
    "from sklearn.preprocessing import OrdinalEncoder, StandardScaler\n",
    "import cudf\n",
    "import cupy\n",
    "# import for model building\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from cuml.metrics.regression import r2_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn import linear_model\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn import model_selection, datasets\n",
    "from cuml.dask.common import utils as dask_utils\n",
    "from dask.distributed import Client, wait\n",
    "from dask_cuda import LocalCUDACluster\n",
    "import dask_cudf\n",
    "from cuml.dask.ensemble import RandomForestClassifier as cumlDaskRF\n",
    "from sklearn.ensemble import RandomForestClassifier as sklRF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll read the dataframe into y from the csv file, view its dimensions and observe the first 5 rows of the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(72, 2)\n",
      "CPU times: user 2.61 ms, sys: 305 Âµs, total: 2.92 ms\n",
      "Wall time: 2.49 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>ALL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient cancer\n",
       "0        1    ALL\n",
       "1        2    ALL\n",
       "2        3    ALL\n",
       "3        4    ALL\n",
       "4        5    ALL"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "y = pd.read_csv('../../../data/actual.csv')\n",
    "print(y.shape)\n",
    "y.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's convert our target variable categories to numbers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "y['cancer'].value_counts()\n",
    "# Recode label to numeric\n",
    "y = y.replace({'ALL':0,'AML':1})\n",
    "labels = ['ALL', 'AML'] # for plotting convenience later on"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the training and test data provided in the challenge from the data folder. View their dimensions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7129, 78)\n",
      "(7129, 70)\n"
     ]
    }
   ],
   "source": [
    "# Import training data\n",
    "df_train = pd.read_csv('../../../data/data_set_ALL_AML_train.csv')\n",
    "print(df_train.shape)\n",
    "\n",
    "# Import testing data\n",
    "df_test = pd.read_csv('../../../data/data_set_ALL_AML_independent.csv')\n",
    "print(df_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe the first few rows of the train dataframe and the data format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>1</th>\n",
       "      <th>call</th>\n",
       "      <th>2</th>\n",
       "      <th>call.1</th>\n",
       "      <th>3</th>\n",
       "      <th>call.2</th>\n",
       "      <th>4</th>\n",
       "      <th>call.3</th>\n",
       "      <th>...</th>\n",
       "      <th>29</th>\n",
       "      <th>call.33</th>\n",
       "      <th>30</th>\n",
       "      <th>call.34</th>\n",
       "      <th>31</th>\n",
       "      <th>call.35</th>\n",
       "      <th>32</th>\n",
       "      <th>call.36</th>\n",
       "      <th>33</th>\n",
       "      <th>call.37</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-214</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-76</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>A</td>\n",
       "      <td>-318</td>\n",
       "      <td>A</td>\n",
       "      <td>-32</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-135</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-73</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-114</td>\n",
       "      <td>A</td>\n",
       "      <td>-192</td>\n",
       "      <td>A</td>\n",
       "      <td>-49</td>\n",
       "      <td>A</td>\n",
       "      <td>-79</td>\n",
       "      <td>A</td>\n",
       "      <td>-186</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>A</td>\n",
       "      <td>-307</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>-95</td>\n",
       "      <td>A</td>\n",
       "      <td>49</td>\n",
       "      <td>A</td>\n",
       "      <td>-37</td>\n",
       "      <td>A</td>\n",
       "      <td>-70</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>88</td>\n",
       "      <td>A</td>\n",
       "      <td>283</td>\n",
       "      <td>A</td>\n",
       "      <td>309</td>\n",
       "      <td>A</td>\n",
       "      <td>12</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>193</td>\n",
       "      <td>A</td>\n",
       "      <td>312</td>\n",
       "      <td>A</td>\n",
       "      <td>230</td>\n",
       "      <td>P</td>\n",
       "      <td>330</td>\n",
       "      <td>A</td>\n",
       "      <td>337</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-295</td>\n",
       "      <td>A</td>\n",
       "      <td>-264</td>\n",
       "      <td>A</td>\n",
       "      <td>-376</td>\n",
       "      <td>A</td>\n",
       "      <td>-419</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-51</td>\n",
       "      <td>A</td>\n",
       "      <td>-139</td>\n",
       "      <td>A</td>\n",
       "      <td>-367</td>\n",
       "      <td>A</td>\n",
       "      <td>-188</td>\n",
       "      <td>A</td>\n",
       "      <td>-407</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene Description Gene Accession Number    1 call    2  \\\n",
       "0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -214    A -139   \n",
       "1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -153    A  -73   \n",
       "2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at  -58    A   -1   \n",
       "3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at   88    A  283   \n",
       "4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -295    A -264   \n",
       "\n",
       "  call.1    3 call.2    4 call.3  ...   29 call.33   30 call.34   31 call.35  \\\n",
       "0      A  -76      A -135      A  ...   15       A -318       A  -32       A   \n",
       "1      A  -49      A -114      A  ... -114       A -192       A  -49       A   \n",
       "2      A -307      A  265      A  ...    2       A  -95       A   49       A   \n",
       "3      A  309      A   12      A  ...  193       A  312       A  230       P   \n",
       "4      A -376      A -419      A  ...  -51       A -139       A -367       A   \n",
       "\n",
       "    32 call.36   33 call.37  \n",
       "0 -124       A -135       A  \n",
       "1  -79       A -186       A  \n",
       "2  -37       A  -70       A  \n",
       "3  330       A  337       A  \n",
       "4 -188       A -407       A  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observe the first few rows of the test dataframe and the data format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Description</th>\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>39</th>\n",
       "      <th>call</th>\n",
       "      <th>40</th>\n",
       "      <th>call.1</th>\n",
       "      <th>42</th>\n",
       "      <th>call.2</th>\n",
       "      <th>47</th>\n",
       "      <th>call.3</th>\n",
       "      <th>...</th>\n",
       "      <th>65</th>\n",
       "      <th>call.29</th>\n",
       "      <th>66</th>\n",
       "      <th>call.30</th>\n",
       "      <th>63</th>\n",
       "      <th>call.31</th>\n",
       "      <th>64</th>\n",
       "      <th>call.32</th>\n",
       "      <th>62</th>\n",
       "      <th>call.33</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>-342</td>\n",
       "      <td>A</td>\n",
       "      <td>-87</td>\n",
       "      <td>A</td>\n",
       "      <td>22</td>\n",
       "      <td>A</td>\n",
       "      <td>-243</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-62</td>\n",
       "      <td>A</td>\n",
       "      <td>-58</td>\n",
       "      <td>A</td>\n",
       "      <td>-161</td>\n",
       "      <td>A</td>\n",
       "      <td>-48</td>\n",
       "      <td>A</td>\n",
       "      <td>-176</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>-200</td>\n",
       "      <td>A</td>\n",
       "      <td>-248</td>\n",
       "      <td>A</td>\n",
       "      <td>-153</td>\n",
       "      <td>A</td>\n",
       "      <td>-218</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-198</td>\n",
       "      <td>A</td>\n",
       "      <td>-217</td>\n",
       "      <td>A</td>\n",
       "      <td>-215</td>\n",
       "      <td>A</td>\n",
       "      <td>-531</td>\n",
       "      <td>A</td>\n",
       "      <td>-284</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>41</td>\n",
       "      <td>A</td>\n",
       "      <td>262</td>\n",
       "      <td>A</td>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>-163</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-5</td>\n",
       "      <td>A</td>\n",
       "      <td>63</td>\n",
       "      <td>A</td>\n",
       "      <td>-46</td>\n",
       "      <td>A</td>\n",
       "      <td>-124</td>\n",
       "      <td>A</td>\n",
       "      <td>-81</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>328</td>\n",
       "      <td>A</td>\n",
       "      <td>295</td>\n",
       "      <td>A</td>\n",
       "      <td>276</td>\n",
       "      <td>A</td>\n",
       "      <td>182</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>141</td>\n",
       "      <td>A</td>\n",
       "      <td>95</td>\n",
       "      <td>A</td>\n",
       "      <td>146</td>\n",
       "      <td>A</td>\n",
       "      <td>431</td>\n",
       "      <td>A</td>\n",
       "      <td>9</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>-224</td>\n",
       "      <td>A</td>\n",
       "      <td>-226</td>\n",
       "      <td>A</td>\n",
       "      <td>-211</td>\n",
       "      <td>A</td>\n",
       "      <td>-289</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-256</td>\n",
       "      <td>A</td>\n",
       "      <td>-191</td>\n",
       "      <td>A</td>\n",
       "      <td>-172</td>\n",
       "      <td>A</td>\n",
       "      <td>-496</td>\n",
       "      <td>A</td>\n",
       "      <td>-294</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Gene Description Gene Accession Number   39 call   40  \\\n",
       "0  AFFX-BioB-5_at (endogenous control)        AFFX-BioB-5_at -342    A  -87   \n",
       "1  AFFX-BioB-M_at (endogenous control)        AFFX-BioB-M_at -200    A -248   \n",
       "2  AFFX-BioB-3_at (endogenous control)        AFFX-BioB-3_at   41    A  262   \n",
       "3  AFFX-BioC-5_at (endogenous control)        AFFX-BioC-5_at  328    A  295   \n",
       "4  AFFX-BioC-3_at (endogenous control)        AFFX-BioC-3_at -224    A -226   \n",
       "\n",
       "  call.1   42 call.2   47 call.3  ...   65 call.29   66 call.30   63 call.31  \\\n",
       "0      A   22      A -243      A  ...  -62       A  -58       A -161       A   \n",
       "1      A -153      A -218      A  ... -198       A -217       A -215       A   \n",
       "2      A   17      A -163      A  ...   -5       A   63       A  -46       A   \n",
       "3      A  276      A  182      A  ...  141       A   95       A  146       A   \n",
       "4      A -211      A -289      A  ... -256       A -191       A -172       A   \n",
       "\n",
       "    64 call.32   62 call.33  \n",
       "0  -48       A -176       A  \n",
       "1 -531       A -284       A  \n",
       "2 -124       A  -81       A  \n",
       "3  431       A    9       A  \n",
       "4 -496       A -294       A  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we can see, the data set has categorical values but only for the columns starting with \"call\". We won't use the columns having categorical values, but remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove \"call\" columns from training and testing data\n",
    "train_to_keep = [col for col in df_train.columns if \"call\" not in col]\n",
    "test_to_keep = [col for col in df_test.columns if \"call\" not in col]\n",
    "\n",
    "X_train_tr = df_train[train_to_keep]\n",
    "X_test_tr = df_test[test_to_keep]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename the columns and reindex for formatting purposes and ease in reading the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_columns_titles = ['Gene Description', 'Gene Accession Number', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',\n",
    "       '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', \n",
    "       '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38']\n",
    "\n",
    "X_train_tr = X_train_tr.reindex(columns=train_columns_titles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_columns_titles = ['Gene Description', 'Gene Accession Number','39', '40', '41', '42', '43', '44', '45', '46',\n",
    "       '47', '48', '49', '50', '51', '52', '53',  '54', '55', '56', '57', '58', '59',\n",
    "       '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72']\n",
    "\n",
    "X_test_tr = X_test_tr.reindex(columns=test_columns_titles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will take the transpose of the dataframe so that each row is a patient and each column is a gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(40, 7129)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>7119</th>\n",
       "      <th>7120</th>\n",
       "      <th>7121</th>\n",
       "      <th>7122</th>\n",
       "      <th>7123</th>\n",
       "      <th>7124</th>\n",
       "      <th>7125</th>\n",
       "      <th>7126</th>\n",
       "      <th>7127</th>\n",
       "      <th>7128</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene Description</th>\n",
       "      <td>AFFX-BioB-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-M_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioC-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioDn-5_at (endogenous control)</td>\n",
       "      <td>AFFX-BioDn-3_at (endogenous control)</td>\n",
       "      <td>AFFX-CreX-5_at (endogenous control)</td>\n",
       "      <td>AFFX-CreX-3_at (endogenous control)</td>\n",
       "      <td>AFFX-BioB-5_st (endogenous control)</td>\n",
       "      <td>...</td>\n",
       "      <td>Transcription factor Stat5b (stat5b) mRNA</td>\n",
       "      <td>Breast epithelial antigen BA46 mRNA</td>\n",
       "      <td>GB DEF = Calcium/calmodulin-dependent protein ...</td>\n",
       "      <td>TUBULIN ALPHA-4 CHAIN</td>\n",
       "      <td>CYP4B1 Cytochrome P450; subfamily IVB; polypep...</td>\n",
       "      <td>PTGER3 Prostaglandin E receptor 3 (subtype EP3...</td>\n",
       "      <td>HMG2 High-mobility group (nonhistone chromosom...</td>\n",
       "      <td>RB1 Retinoblastoma 1 (including osteosarcoma)</td>\n",
       "      <td>GB DEF = Glycophorin Sta (type A) exons 3 and ...</td>\n",
       "      <td>GB DEF = mRNA (clone 1A7)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Accession Number</th>\n",
       "      <td>AFFX-BioB-5_at</td>\n",
       "      <td>AFFX-BioB-M_at</td>\n",
       "      <td>AFFX-BioB-3_at</td>\n",
       "      <td>AFFX-BioC-5_at</td>\n",
       "      <td>AFFX-BioC-3_at</td>\n",
       "      <td>AFFX-BioDn-5_at</td>\n",
       "      <td>AFFX-BioDn-3_at</td>\n",
       "      <td>AFFX-CreX-5_at</td>\n",
       "      <td>AFFX-CreX-3_at</td>\n",
       "      <td>AFFX-BioB-5_st</td>\n",
       "      <td>...</td>\n",
       "      <td>U48730_at</td>\n",
       "      <td>U58516_at</td>\n",
       "      <td>U73738_at</td>\n",
       "      <td>X06956_at</td>\n",
       "      <td>X16699_at</td>\n",
       "      <td>X83863_at</td>\n",
       "      <td>Z17240_at</td>\n",
       "      <td>L49218_f_at</td>\n",
       "      <td>M71243_f_at</td>\n",
       "      <td>Z78285_f_at</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-214</td>\n",
       "      <td>-153</td>\n",
       "      <td>-58</td>\n",
       "      <td>88</td>\n",
       "      <td>-295</td>\n",
       "      <td>-558</td>\n",
       "      <td>199</td>\n",
       "      <td>-176</td>\n",
       "      <td>252</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>511</td>\n",
       "      <td>-125</td>\n",
       "      <td>389</td>\n",
       "      <td>-37</td>\n",
       "      <td>793</td>\n",
       "      <td>329</td>\n",
       "      <td>36</td>\n",
       "      <td>191</td>\n",
       "      <td>-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-139</td>\n",
       "      <td>-73</td>\n",
       "      <td>-1</td>\n",
       "      <td>283</td>\n",
       "      <td>-264</td>\n",
       "      <td>-400</td>\n",
       "      <td>-330</td>\n",
       "      <td>-168</td>\n",
       "      <td>101</td>\n",
       "      <td>74</td>\n",
       "      <td>...</td>\n",
       "      <td>169</td>\n",
       "      <td>837</td>\n",
       "      <td>-36</td>\n",
       "      <td>442</td>\n",
       "      <td>-17</td>\n",
       "      <td>782</td>\n",
       "      <td>295</td>\n",
       "      <td>11</td>\n",
       "      <td>76</td>\n",
       "      <td>-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-76</td>\n",
       "      <td>-49</td>\n",
       "      <td>-307</td>\n",
       "      <td>309</td>\n",
       "      <td>-376</td>\n",
       "      <td>-650</td>\n",
       "      <td>33</td>\n",
       "      <td>-367</td>\n",
       "      <td>206</td>\n",
       "      <td>-215</td>\n",
       "      <td>...</td>\n",
       "      <td>315</td>\n",
       "      <td>1199</td>\n",
       "      <td>33</td>\n",
       "      <td>168</td>\n",
       "      <td>52</td>\n",
       "      <td>1138</td>\n",
       "      <td>777</td>\n",
       "      <td>41</td>\n",
       "      <td>228</td>\n",
       "      <td>-41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      0     \\\n",
       "Gene Description       AFFX-BioB-5_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioB-5_at   \n",
       "1                                                     -214   \n",
       "2                                                     -139   \n",
       "3                                                      -76   \n",
       "\n",
       "                                                      1     \\\n",
       "Gene Description       AFFX-BioB-M_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioB-M_at   \n",
       "1                                                     -153   \n",
       "2                                                      -73   \n",
       "3                                                      -49   \n",
       "\n",
       "                                                      2     \\\n",
       "Gene Description       AFFX-BioB-3_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioB-3_at   \n",
       "1                                                      -58   \n",
       "2                                                       -1   \n",
       "3                                                     -307   \n",
       "\n",
       "                                                      3     \\\n",
       "Gene Description       AFFX-BioC-5_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioC-5_at   \n",
       "1                                                       88   \n",
       "2                                                      283   \n",
       "3                                                      309   \n",
       "\n",
       "                                                      4     \\\n",
       "Gene Description       AFFX-BioC-3_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioC-3_at   \n",
       "1                                                     -295   \n",
       "2                                                     -264   \n",
       "3                                                     -376   \n",
       "\n",
       "                                                       5     \\\n",
       "Gene Description       AFFX-BioDn-5_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioDn-5_at   \n",
       "1                                                      -558   \n",
       "2                                                      -400   \n",
       "3                                                      -650   \n",
       "\n",
       "                                                       6     \\\n",
       "Gene Description       AFFX-BioDn-3_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-BioDn-3_at   \n",
       "1                                                       199   \n",
       "2                                                      -330   \n",
       "3                                                        33   \n",
       "\n",
       "                                                      7     \\\n",
       "Gene Description       AFFX-CreX-5_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-CreX-5_at   \n",
       "1                                                     -176   \n",
       "2                                                     -168   \n",
       "3                                                     -367   \n",
       "\n",
       "                                                      8     \\\n",
       "Gene Description       AFFX-CreX-3_at (endogenous control)   \n",
       "Gene Accession Number                       AFFX-CreX-3_at   \n",
       "1                                                      252   \n",
       "2                                                      101   \n",
       "3                                                      206   \n",
       "\n",
       "                                                      9     ...  \\\n",
       "Gene Description       AFFX-BioB-5_st (endogenous control)  ...   \n",
       "Gene Accession Number                       AFFX-BioB-5_st  ...   \n",
       "1                                                      206  ...   \n",
       "2                                                       74  ...   \n",
       "3                                                     -215  ...   \n",
       "\n",
       "                                                            7119  \\\n",
       "Gene Description       Transcription factor Stat5b (stat5b) mRNA   \n",
       "Gene Accession Number                                  U48730_at   \n",
       "1                                                            185   \n",
       "2                                                            169   \n",
       "3                                                            315   \n",
       "\n",
       "                                                      7120  \\\n",
       "Gene Description       Breast epithelial antigen BA46 mRNA   \n",
       "Gene Accession Number                            U58516_at   \n",
       "1                                                      511   \n",
       "2                                                      837   \n",
       "3                                                     1199   \n",
       "\n",
       "                                                                    7121  \\\n",
       "Gene Description       GB DEF = Calcium/calmodulin-dependent protein ...   \n",
       "Gene Accession Number                                          U73738_at   \n",
       "1                                                                   -125   \n",
       "2                                                                    -36   \n",
       "3                                                                     33   \n",
       "\n",
       "                                        7122  \\\n",
       "Gene Description       TUBULIN ALPHA-4 CHAIN   \n",
       "Gene Accession Number              X06956_at   \n",
       "1                                        389   \n",
       "2                                        442   \n",
       "3                                        168   \n",
       "\n",
       "                                                                    7123  \\\n",
       "Gene Description       CYP4B1 Cytochrome P450; subfamily IVB; polypep...   \n",
       "Gene Accession Number                                          X16699_at   \n",
       "1                                                                    -37   \n",
       "2                                                                    -17   \n",
       "3                                                                     52   \n",
       "\n",
       "                                                                    7124  \\\n",
       "Gene Description       PTGER3 Prostaglandin E receptor 3 (subtype EP3...   \n",
       "Gene Accession Number                                          X83863_at   \n",
       "1                                                                    793   \n",
       "2                                                                    782   \n",
       "3                                                                   1138   \n",
       "\n",
       "                                                                    7125  \\\n",
       "Gene Description       HMG2 High-mobility group (nonhistone chromosom...   \n",
       "Gene Accession Number                                          Z17240_at   \n",
       "1                                                                    329   \n",
       "2                                                                    295   \n",
       "3                                                                    777   \n",
       "\n",
       "                                                                7126  \\\n",
       "Gene Description       RB1 Retinoblastoma 1 (including osteosarcoma)   \n",
       "Gene Accession Number                                    L49218_f_at   \n",
       "1                                                                 36   \n",
       "2                                                                 11   \n",
       "3                                                                 41   \n",
       "\n",
       "                                                                    7127  \\\n",
       "Gene Description       GB DEF = Glycophorin Sta (type A) exons 3 and ...   \n",
       "Gene Accession Number                                        M71243_f_at   \n",
       "1                                                                    191   \n",
       "2                                                                     76   \n",
       "3                                                                    228   \n",
       "\n",
       "                                            7128  \n",
       "Gene Description       GB DEF = mRNA (clone 1A7)  \n",
       "Gene Accession Number                Z78285_f_at  \n",
       "1                                            -37  \n",
       "2                                            -14  \n",
       "3                                            -41  \n",
       "\n",
       "[5 rows x 7129 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train = X_train_tr.T\n",
    "X_test = X_test_tr.T\n",
    "\n",
    "print(X_train.shape) \n",
    "X_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just clearning the data, removing extra columns and converting to numerical values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38, 7129)\n",
      "(34, 7129)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <th>AFFX-BioDn-5_at</th>\n",
       "      <th>AFFX-BioDn-3_at</th>\n",
       "      <th>AFFX-CreX-5_at</th>\n",
       "      <th>AFFX-CreX-3_at</th>\n",
       "      <th>AFFX-BioB-5_st</th>\n",
       "      <th>...</th>\n",
       "      <th>U48730_at</th>\n",
       "      <th>U58516_at</th>\n",
       "      <th>U73738_at</th>\n",
       "      <th>X06956_at</th>\n",
       "      <th>X16699_at</th>\n",
       "      <th>X83863_at</th>\n",
       "      <th>Z17240_at</th>\n",
       "      <th>L49218_f_at</th>\n",
       "      <th>M71243_f_at</th>\n",
       "      <th>Z78285_f_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-214</td>\n",
       "      <td>-153</td>\n",
       "      <td>-58</td>\n",
       "      <td>88</td>\n",
       "      <td>-295</td>\n",
       "      <td>-558</td>\n",
       "      <td>199</td>\n",
       "      <td>-176</td>\n",
       "      <td>252</td>\n",
       "      <td>206</td>\n",
       "      <td>...</td>\n",
       "      <td>185</td>\n",
       "      <td>511</td>\n",
       "      <td>-125</td>\n",
       "      <td>389</td>\n",
       "      <td>-37</td>\n",
       "      <td>793</td>\n",
       "      <td>329</td>\n",
       "      <td>36</td>\n",
       "      <td>191</td>\n",
       "      <td>-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-139</td>\n",
       "      <td>-73</td>\n",
       "      <td>-1</td>\n",
       "      <td>283</td>\n",
       "      <td>-264</td>\n",
       "      <td>-400</td>\n",
       "      <td>-330</td>\n",
       "      <td>-168</td>\n",
       "      <td>101</td>\n",
       "      <td>74</td>\n",
       "      <td>...</td>\n",
       "      <td>169</td>\n",
       "      <td>837</td>\n",
       "      <td>-36</td>\n",
       "      <td>442</td>\n",
       "      <td>-17</td>\n",
       "      <td>782</td>\n",
       "      <td>295</td>\n",
       "      <td>11</td>\n",
       "      <td>76</td>\n",
       "      <td>-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-76</td>\n",
       "      <td>-49</td>\n",
       "      <td>-307</td>\n",
       "      <td>309</td>\n",
       "      <td>-376</td>\n",
       "      <td>-650</td>\n",
       "      <td>33</td>\n",
       "      <td>-367</td>\n",
       "      <td>206</td>\n",
       "      <td>-215</td>\n",
       "      <td>...</td>\n",
       "      <td>315</td>\n",
       "      <td>1199</td>\n",
       "      <td>33</td>\n",
       "      <td>168</td>\n",
       "      <td>52</td>\n",
       "      <td>1138</td>\n",
       "      <td>777</td>\n",
       "      <td>41</td>\n",
       "      <td>228</td>\n",
       "      <td>-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-135</td>\n",
       "      <td>-114</td>\n",
       "      <td>265</td>\n",
       "      <td>12</td>\n",
       "      <td>-419</td>\n",
       "      <td>-585</td>\n",
       "      <td>158</td>\n",
       "      <td>-253</td>\n",
       "      <td>49</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>240</td>\n",
       "      <td>835</td>\n",
       "      <td>218</td>\n",
       "      <td>174</td>\n",
       "      <td>-110</td>\n",
       "      <td>627</td>\n",
       "      <td>170</td>\n",
       "      <td>-50</td>\n",
       "      <td>126</td>\n",
       "      <td>-91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-106</td>\n",
       "      <td>-125</td>\n",
       "      <td>-76</td>\n",
       "      <td>168</td>\n",
       "      <td>-230</td>\n",
       "      <td>-284</td>\n",
       "      <td>4</td>\n",
       "      <td>-122</td>\n",
       "      <td>70</td>\n",
       "      <td>252</td>\n",
       "      <td>...</td>\n",
       "      <td>156</td>\n",
       "      <td>649</td>\n",
       "      <td>57</td>\n",
       "      <td>504</td>\n",
       "      <td>-26</td>\n",
       "      <td>250</td>\n",
       "      <td>314</td>\n",
       "      <td>14</td>\n",
       "      <td>56</td>\n",
       "      <td>-25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \\\n",
       "1                                -214            -153             -58   \n",
       "2                                -139             -73              -1   \n",
       "3                                 -76             -49            -307   \n",
       "4                                -135            -114             265   \n",
       "5                                -106            -125             -76   \n",
       "\n",
       "Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \\\n",
       "1                                  88            -295             -558   \n",
       "2                                 283            -264             -400   \n",
       "3                                 309            -376             -650   \n",
       "4                                  12            -419             -585   \n",
       "5                                 168            -230             -284   \n",
       "\n",
       "Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \\\n",
       "1                                  199            -176             252   \n",
       "2                                 -330            -168             101   \n",
       "3                                   33            -367             206   \n",
       "4                                  158            -253              49   \n",
       "5                                    4            -122              70   \n",
       "\n",
       "Gene Accession Number  AFFX-BioB-5_st  ...  U48730_at  U58516_at  U73738_at  \\\n",
       "1                                 206  ...        185        511       -125   \n",
       "2                                  74  ...        169        837        -36   \n",
       "3                                -215  ...        315       1199         33   \n",
       "4                                  31  ...        240        835        218   \n",
       "5                                 252  ...        156        649         57   \n",
       "\n",
       "Gene Accession Number  X06956_at  X16699_at  X83863_at  Z17240_at  \\\n",
       "1                            389        -37        793        329   \n",
       "2                            442        -17        782        295   \n",
       "3                            168         52       1138        777   \n",
       "4                            174       -110        627        170   \n",
       "5                            504        -26        250        314   \n",
       "\n",
       "Gene Accession Number  L49218_f_at  M71243_f_at  Z78285_f_at  \n",
       "1                               36          191          -37  \n",
       "2                               11           76          -14  \n",
       "3                               41          228          -41  \n",
       "4                              -50          126          -91  \n",
       "5                               14           56          -25  \n",
       "\n",
       "[5 rows x 7129 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean up the column names for training and testing data\n",
    "X_train.columns = X_train.iloc[1]\n",
    "X_train = X_train.drop([\"Gene Description\", \"Gene Accession Number\"]).apply(pd.to_numeric)\n",
    "\n",
    "# Clean up the column names for Testing data\n",
    "X_test.columns = X_test.iloc[1]\n",
    "X_test = X_test.drop([\"Gene Description\", \"Gene Accession Number\"]).apply(pd.to_numeric)\n",
    "\n",
    "print(X_train.shape)\n",
    "print(X_test.shape)\n",
    "X_train.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have the 38 patients as rows in the training set, and the other 34 as rows in the testing set. Each of those datasets has 7129 gene expression features. But we haven't yet associated the target labels with the right patients. You will recall that all the labels are all stored in a single dataframe. Let's split the data so that the patients and labels match up across the training and testing dataframes.We are now splitting the data into train and test sets. We will subset the first 38 patient's cancer types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = X_train.reset_index(drop=True)\n",
    "y_train = y[y.patient <= 38].reset_index(drop=True)\n",
    "\n",
    "# Subset the rest for testing\n",
    "X_test = X_test.reset_index(drop=True)\n",
    "y_test = y[y.patient > 38].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generate descriptive statistics to analyse the data further."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene Accession Number</th>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <th>AFFX-BioDn-5_at</th>\n",
       "      <th>AFFX-BioDn-3_at</th>\n",
       "      <th>AFFX-CreX-5_at</th>\n",
       "      <th>AFFX-CreX-3_at</th>\n",
       "      <th>AFFX-BioB-5_st</th>\n",
       "      <th>...</th>\n",
       "      <th>U48730_at</th>\n",
       "      <th>U58516_at</th>\n",
       "      <th>U73738_at</th>\n",
       "      <th>X06956_at</th>\n",
       "      <th>X16699_at</th>\n",
       "      <th>X83863_at</th>\n",
       "      <th>Z17240_at</th>\n",
       "      <th>L49218_f_at</th>\n",
       "      <th>M71243_f_at</th>\n",
       "      <th>Z78285_f_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "      <td>38.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>-120.868421</td>\n",
       "      <td>-150.526316</td>\n",
       "      <td>-17.157895</td>\n",
       "      <td>181.394737</td>\n",
       "      <td>-276.552632</td>\n",
       "      <td>-439.210526</td>\n",
       "      <td>-43.578947</td>\n",
       "      <td>-201.184211</td>\n",
       "      <td>99.052632</td>\n",
       "      <td>112.131579</td>\n",
       "      <td>...</td>\n",
       "      <td>178.763158</td>\n",
       "      <td>750.842105</td>\n",
       "      <td>8.815789</td>\n",
       "      <td>399.131579</td>\n",
       "      <td>-20.052632</td>\n",
       "      <td>869.052632</td>\n",
       "      <td>335.842105</td>\n",
       "      <td>19.210526</td>\n",
       "      <td>504.394737</td>\n",
       "      <td>-29.210526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>109.555656</td>\n",
       "      <td>75.734507</td>\n",
       "      <td>117.686144</td>\n",
       "      <td>117.468004</td>\n",
       "      <td>111.004431</td>\n",
       "      <td>135.458412</td>\n",
       "      <td>219.482393</td>\n",
       "      <td>90.838989</td>\n",
       "      <td>83.178397</td>\n",
       "      <td>211.815597</td>\n",
       "      <td>...</td>\n",
       "      <td>84.826830</td>\n",
       "      <td>298.008392</td>\n",
       "      <td>77.108507</td>\n",
       "      <td>469.579868</td>\n",
       "      <td>42.346031</td>\n",
       "      <td>482.366461</td>\n",
       "      <td>209.826766</td>\n",
       "      <td>31.158841</td>\n",
       "      <td>728.744405</td>\n",
       "      <td>30.851132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-476.000000</td>\n",
       "      <td>-327.000000</td>\n",
       "      <td>-307.000000</td>\n",
       "      <td>-36.000000</td>\n",
       "      <td>-541.000000</td>\n",
       "      <td>-790.000000</td>\n",
       "      <td>-479.000000</td>\n",
       "      <td>-463.000000</td>\n",
       "      <td>-82.000000</td>\n",
       "      <td>-215.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>30.000000</td>\n",
       "      <td>224.000000</td>\n",
       "      <td>-178.000000</td>\n",
       "      <td>36.000000</td>\n",
       "      <td>-112.000000</td>\n",
       "      <td>195.000000</td>\n",
       "      <td>41.000000</td>\n",
       "      <td>-50.000000</td>\n",
       "      <td>-2.000000</td>\n",
       "      <td>-94.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-138.750000</td>\n",
       "      <td>-205.000000</td>\n",
       "      <td>-83.250000</td>\n",
       "      <td>81.250000</td>\n",
       "      <td>-374.250000</td>\n",
       "      <td>-547.000000</td>\n",
       "      <td>-169.000000</td>\n",
       "      <td>-239.250000</td>\n",
       "      <td>36.000000</td>\n",
       "      <td>-47.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>120.000000</td>\n",
       "      <td>575.500000</td>\n",
       "      <td>-42.750000</td>\n",
       "      <td>174.500000</td>\n",
       "      <td>-48.000000</td>\n",
       "      <td>595.250000</td>\n",
       "      <td>232.750000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>136.000000</td>\n",
       "      <td>-42.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>-106.500000</td>\n",
       "      <td>-141.500000</td>\n",
       "      <td>-43.500000</td>\n",
       "      <td>200.000000</td>\n",
       "      <td>-263.000000</td>\n",
       "      <td>-426.500000</td>\n",
       "      <td>-33.500000</td>\n",
       "      <td>-185.500000</td>\n",
       "      <td>99.500000</td>\n",
       "      <td>70.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>174.500000</td>\n",
       "      <td>700.000000</td>\n",
       "      <td>10.500000</td>\n",
       "      <td>266.000000</td>\n",
       "      <td>-18.000000</td>\n",
       "      <td>744.500000</td>\n",
       "      <td>308.500000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>243.500000</td>\n",
       "      <td>-26.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>-68.250000</td>\n",
       "      <td>-94.750000</td>\n",
       "      <td>47.250000</td>\n",
       "      <td>279.250000</td>\n",
       "      <td>-188.750000</td>\n",
       "      <td>-344.750000</td>\n",
       "      <td>79.000000</td>\n",
       "      <td>-144.750000</td>\n",
       "      <td>152.250000</td>\n",
       "      <td>242.750000</td>\n",
       "      <td>...</td>\n",
       "      <td>231.750000</td>\n",
       "      <td>969.500000</td>\n",
       "      <td>57.000000</td>\n",
       "      <td>451.750000</td>\n",
       "      <td>9.250000</td>\n",
       "      <td>1112.000000</td>\n",
       "      <td>389.500000</td>\n",
       "      <td>30.250000</td>\n",
       "      <td>487.250000</td>\n",
       "      <td>-11.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>17.000000</td>\n",
       "      <td>-20.000000</td>\n",
       "      <td>265.000000</td>\n",
       "      <td>392.000000</td>\n",
       "      <td>-51.000000</td>\n",
       "      <td>-155.000000</td>\n",
       "      <td>419.000000</td>\n",
       "      <td>-24.000000</td>\n",
       "      <td>283.000000</td>\n",
       "      <td>561.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>356.000000</td>\n",
       "      <td>1653.000000</td>\n",
       "      <td>218.000000</td>\n",
       "      <td>2527.000000</td>\n",
       "      <td>52.000000</td>\n",
       "      <td>2315.000000</td>\n",
       "      <td>1109.000000</td>\n",
       "      <td>115.000000</td>\n",
       "      <td>3193.000000</td>\n",
       "      <td>36.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene Accession Number  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioB-3_at  \\\n",
       "count                       38.000000       38.000000       38.000000   \n",
       "mean                      -120.868421     -150.526316      -17.157895   \n",
       "std                        109.555656       75.734507      117.686144   \n",
       "min                       -476.000000     -327.000000     -307.000000   \n",
       "25%                       -138.750000     -205.000000      -83.250000   \n",
       "50%                       -106.500000     -141.500000      -43.500000   \n",
       "75%                        -68.250000      -94.750000       47.250000   \n",
       "max                         17.000000      -20.000000      265.000000   \n",
       "\n",
       "Gene Accession Number  AFFX-BioC-5_at  AFFX-BioC-3_at  AFFX-BioDn-5_at  \\\n",
       "count                       38.000000       38.000000        38.000000   \n",
       "mean                       181.394737     -276.552632      -439.210526   \n",
       "std                        117.468004      111.004431       135.458412   \n",
       "min                        -36.000000     -541.000000      -790.000000   \n",
       "25%                         81.250000     -374.250000      -547.000000   \n",
       "50%                        200.000000     -263.000000      -426.500000   \n",
       "75%                        279.250000     -188.750000      -344.750000   \n",
       "max                        392.000000      -51.000000      -155.000000   \n",
       "\n",
       "Gene Accession Number  AFFX-BioDn-3_at  AFFX-CreX-5_at  AFFX-CreX-3_at  \\\n",
       "count                        38.000000       38.000000       38.000000   \n",
       "mean                        -43.578947     -201.184211       99.052632   \n",
       "std                         219.482393       90.838989       83.178397   \n",
       "min                        -479.000000     -463.000000      -82.000000   \n",
       "25%                        -169.000000     -239.250000       36.000000   \n",
       "50%                         -33.500000     -185.500000       99.500000   \n",
       "75%                          79.000000     -144.750000      152.250000   \n",
       "max                         419.000000      -24.000000      283.000000   \n",
       "\n",
       "Gene Accession Number  AFFX-BioB-5_st  ...   U48730_at    U58516_at  \\\n",
       "count                       38.000000  ...   38.000000    38.000000   \n",
       "mean                       112.131579  ...  178.763158   750.842105   \n",
       "std                        211.815597  ...   84.826830   298.008392   \n",
       "min                       -215.000000  ...   30.000000   224.000000   \n",
       "25%                        -47.000000  ...  120.000000   575.500000   \n",
       "50%                         70.500000  ...  174.500000   700.000000   \n",
       "75%                        242.750000  ...  231.750000   969.500000   \n",
       "max                        561.000000  ...  356.000000  1653.000000   \n",
       "\n",
       "Gene Accession Number   U73738_at    X06956_at   X16699_at    X83863_at  \\\n",
       "count                   38.000000    38.000000   38.000000    38.000000   \n",
       "mean                     8.815789   399.131579  -20.052632   869.052632   \n",
       "std                     77.108507   469.579868   42.346031   482.366461   \n",
       "min                   -178.000000    36.000000 -112.000000   195.000000   \n",
       "25%                    -42.750000   174.500000  -48.000000   595.250000   \n",
       "50%                     10.500000   266.000000  -18.000000   744.500000   \n",
       "75%                     57.000000   451.750000    9.250000  1112.000000   \n",
       "max                    218.000000  2527.000000   52.000000  2315.000000   \n",
       "\n",
       "Gene Accession Number    Z17240_at  L49218_f_at  M71243_f_at  Z78285_f_at  \n",
       "count                    38.000000    38.000000    38.000000    38.000000  \n",
       "mean                    335.842105    19.210526   504.394737   -29.210526  \n",
       "std                     209.826766    31.158841   728.744405    30.851132  \n",
       "min                      41.000000   -50.000000    -2.000000   -94.000000  \n",
       "25%                     232.750000     8.000000   136.000000   -42.750000  \n",
       "50%                     308.500000    20.000000   243.500000   -26.000000  \n",
       "75%                     389.500000    30.250000   487.250000   -11.500000  \n",
       "max                    1109.000000   115.000000  3193.000000    36.000000  \n",
       "\n",
       "[8 rows x 7129 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clearly there is some variation in the scales across the different features. Many machine learning models work much better with data that's on the same scale, so let's create a scaled version of the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_fl = X_train.astype(float, 64)\n",
    "X_test_fl = X_test.astype(float, 64)\n",
    "\n",
    "# Apply the same scaling to both datasets\n",
    "scaler = StandardScaler()\n",
    "X_train = scaler.fit_transform(X_train_fl)\n",
    "X_test = scaler.transform(X_test_fl) # note that we transform rather than fit_transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>56</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>59</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>61</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>62</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>63</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>65</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>67</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>70</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    patient  cancer\n",
       "0        39       0\n",
       "1        40       0\n",
       "2        41       0\n",
       "3        42       0\n",
       "4        43       0\n",
       "5        44       0\n",
       "6        45       0\n",
       "7        46       0\n",
       "8        47       0\n",
       "9        48       0\n",
       "10       49       0\n",
       "11       50       1\n",
       "12       51       1\n",
       "13       52       1\n",
       "14       53       1\n",
       "15       54       1\n",
       "16       55       0\n",
       "17       56       0\n",
       "18       57       1\n",
       "19       58       1\n",
       "20       59       0\n",
       "21       60       1\n",
       "22       61       1\n",
       "23       62       1\n",
       "24       63       1\n",
       "25       64       1\n",
       "26       65       1\n",
       "27       66       1\n",
       "28       67       0\n",
       "29       68       0\n",
       "30       69       0\n",
       "31       70       0\n",
       "32       71       0\n",
       "33       72       0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='dask1'></a>\n",
    "\n",
    "### 2. Conversion to CuDF Dataframe\n",
    "Convert the pandas dataframes to CuDF dataframes to carry out the further CuML tasks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Modify the code in this cell\n",
    "\n",
    "#%%time\n",
    "X_cudf_train = cudf.DataFrame(X_train)  #Pass X train dataframe here\n",
    "X_cudf_test = cudf.DataFrame(X_test)    #Pass X test dataframe here\n",
    "\n",
    "y_cudf_train = cudf.DataFrame(y_train['cancer'].values)  #Pass y train dataframe here\n",
    "y_cudf_test = cudf.Series(y_test['cancer'].values)   #Pass y test dataframe here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>7119</th>\n",
       "      <th>7120</th>\n",
       "      <th>7121</th>\n",
       "      <th>7122</th>\n",
       "      <th>7123</th>\n",
       "      <th>7124</th>\n",
       "      <th>7125</th>\n",
       "      <th>7126</th>\n",
       "      <th>7127</th>\n",
       "      <th>7128</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.861496</td>\n",
       "      <td>-0.033101</td>\n",
       "      <td>-0.351701</td>\n",
       "      <td>-0.805738</td>\n",
       "      <td>-0.168417</td>\n",
       "      <td>-0.888716</td>\n",
       "      <td>1.120068</td>\n",
       "      <td>0.280962</td>\n",
       "      <td>1.863470</td>\n",
       "      <td>0.449110</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074511</td>\n",
       "      <td>-0.815620</td>\n",
       "      <td>-1.758717</td>\n",
       "      <td>-0.021865</td>\n",
       "      <td>-0.405584</td>\n",
       "      <td>-0.159782</td>\n",
       "      <td>-0.033046</td>\n",
       "      <td>0.546068</td>\n",
       "      <td>-0.435820</td>\n",
       "      <td>-0.255875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.167723</td>\n",
       "      <td>1.037400</td>\n",
       "      <td>0.139139</td>\n",
       "      <td>0.876572</td>\n",
       "      <td>0.114600</td>\n",
       "      <td>0.293351</td>\n",
       "      <td>-1.322502</td>\n",
       "      <td>0.370212</td>\n",
       "      <td>0.023726</td>\n",
       "      <td>-0.182439</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.116640</td>\n",
       "      <td>0.292993</td>\n",
       "      <td>-0.589006</td>\n",
       "      <td>0.092516</td>\n",
       "      <td>0.073055</td>\n",
       "      <td>-0.182892</td>\n",
       "      <td>-0.197260</td>\n",
       "      <td>-0.267043</td>\n",
       "      <td>-0.595744</td>\n",
       "      <td>0.499648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.415047</td>\n",
       "      <td>1.358550</td>\n",
       "      <td>-2.495899</td>\n",
       "      <td>1.100880</td>\n",
       "      <td>-0.907912</td>\n",
       "      <td>-1.577008</td>\n",
       "      <td>0.353591</td>\n",
       "      <td>-1.849884</td>\n",
       "      <td>1.303018</td>\n",
       "      <td>-1.565148</td>\n",
       "      <td>...</td>\n",
       "      <td>1.627617</td>\n",
       "      <td>1.524030</td>\n",
       "      <td>0.317849</td>\n",
       "      <td>-0.498816</td>\n",
       "      <td>1.724361</td>\n",
       "      <td>0.565043</td>\n",
       "      <td>2.130709</td>\n",
       "      <td>0.708690</td>\n",
       "      <td>-0.384366</td>\n",
       "      <td>-0.387270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.130721</td>\n",
       "      <td>0.488768</td>\n",
       "      <td>2.429729</td>\n",
       "      <td>-1.461407</td>\n",
       "      <td>-1.300484</td>\n",
       "      <td>-1.090715</td>\n",
       "      <td>0.930757</td>\n",
       "      <td>-0.578070</td>\n",
       "      <td>-0.609828</td>\n",
       "      <td>-0.388171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.731595</td>\n",
       "      <td>0.286192</td>\n",
       "      <td>2.749271</td>\n",
       "      <td>-0.485868</td>\n",
       "      <td>-2.152617</td>\n",
       "      <td>-0.508538</td>\n",
       "      <td>-0.800986</td>\n",
       "      <td>-2.251033</td>\n",
       "      <td>-0.526212</td>\n",
       "      <td>-2.029712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.137537</td>\n",
       "      <td>0.341574</td>\n",
       "      <td>-0.506703</td>\n",
       "      <td>-0.115559</td>\n",
       "      <td>0.425006</td>\n",
       "      <td>1.161198</td>\n",
       "      <td>0.219688</td>\n",
       "      <td>0.883400</td>\n",
       "      <td>-0.353970</td>\n",
       "      <td>0.669195</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.271951</td>\n",
       "      <td>-0.346330</td>\n",
       "      <td>0.633277</td>\n",
       "      <td>0.226322</td>\n",
       "      <td>-0.142332</td>\n",
       "      <td>-1.300593</td>\n",
       "      <td>-0.105493</td>\n",
       "      <td>-0.169469</td>\n",
       "      <td>-0.623557</td>\n",
       "      <td>0.138311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-0.158472</td>\n",
       "      <td>0.876825</td>\n",
       "      <td>1.999167</td>\n",
       "      <td>-0.952401</td>\n",
       "      <td>0.041564</td>\n",
       "      <td>-0.888716</td>\n",
       "      <td>0.510580</td>\n",
       "      <td>0.169399</td>\n",
       "      <td>-0.146846</td>\n",
       "      <td>0.386912</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.761776</td>\n",
       "      <td>1.598844</td>\n",
       "      <td>-1.114719</td>\n",
       "      <td>-0.490184</td>\n",
       "      <td>-1.291066</td>\n",
       "      <td>-0.470721</td>\n",
       "      <td>0.024912</td>\n",
       "      <td>0.220824</td>\n",
       "      <td>-0.433039</td>\n",
       "      <td>-0.781456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.452048</td>\n",
       "      <td>0.087330</td>\n",
       "      <td>2.197225</td>\n",
       "      <td>-1.090436</td>\n",
       "      <td>-1.117893</td>\n",
       "      <td>-0.836346</td>\n",
       "      <td>0.806089</td>\n",
       "      <td>0.247493</td>\n",
       "      <td>0.328320</td>\n",
       "      <td>-0.632178</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.777269</td>\n",
       "      <td>0.231781</td>\n",
       "      <td>-2.455287</td>\n",
       "      <td>-0.535505</td>\n",
       "      <td>0.049123</td>\n",
       "      <td>0.569244</td>\n",
       "      <td>0.705915</td>\n",
       "      <td>-0.299567</td>\n",
       "      <td>-0.188286</td>\n",
       "      <td>-0.420119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-2.702307</td>\n",
       "      <td>-1.464896</td>\n",
       "      <td>0.208029</td>\n",
       "      <td>-1.582188</td>\n",
       "      <td>-2.414292</td>\n",
       "      <td>-2.624409</td>\n",
       "      <td>-1.068548</td>\n",
       "      <td>-2.920885</td>\n",
       "      <td>-0.353970</td>\n",
       "      <td>-1.345063</td>\n",
       "      <td>...</td>\n",
       "      <td>1.316996</td>\n",
       "      <td>-0.414343</td>\n",
       "      <td>-1.246147</td>\n",
       "      <td>-0.209624</td>\n",
       "      <td>1.030334</td>\n",
       "      <td>1.953764</td>\n",
       "      <td>0.532042</td>\n",
       "      <td>1.294130</td>\n",
       "      <td>0.384659</td>\n",
       "      <td>1.616508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1.164322</td>\n",
       "      <td>0.314812</td>\n",
       "      <td>1.060542</td>\n",
       "      <td>0.747163</td>\n",
       "      <td>0.607597</td>\n",
       "      <td>-0.716643</td>\n",
       "      <td>0.201219</td>\n",
       "      <td>0.303274</td>\n",
       "      <td>-0.914421</td>\n",
       "      <td>1.884448</td>\n",
       "      <td>...</td>\n",
       "      <td>2.117443</td>\n",
       "      <td>0.779287</td>\n",
       "      <td>-0.037007</td>\n",
       "      <td>-0.479393</td>\n",
       "      <td>0.192715</td>\n",
       "      <td>-0.233315</td>\n",
       "      <td>0.237423</td>\n",
       "      <td>3.115498</td>\n",
       "      <td>-0.362116</td>\n",
       "      <td>-0.321573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.304043</td>\n",
       "      <td>0.609200</td>\n",
       "      <td>0.509423</td>\n",
       "      <td>0.324429</td>\n",
       "      <td>0.899744</td>\n",
       "      <td>1.445492</td>\n",
       "      <td>1.715704</td>\n",
       "      <td>0.593337</td>\n",
       "      <td>0.949690</td>\n",
       "      <td>0.339067</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.633905</td>\n",
       "      <td>0.799691</td>\n",
       "      <td>0.225849</td>\n",
       "      <td>-0.643412</td>\n",
       "      <td>0.982470</td>\n",
       "      <td>-0.628292</td>\n",
       "      <td>0.111848</td>\n",
       "      <td>-0.332092</td>\n",
       "      <td>-0.463633</td>\n",
       "      <td>1.189473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-0.408231</td>\n",
       "      <td>-0.059864</td>\n",
       "      <td>-0.463647</td>\n",
       "      <td>-0.857501</td>\n",
       "      <td>1.036687</td>\n",
       "      <td>0.068908</td>\n",
       "      <td>0.662952</td>\n",
       "      <td>1.028431</td>\n",
       "      <td>-0.524542</td>\n",
       "      <td>1.138072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074511</td>\n",
       "      <td>-0.370134</td>\n",
       "      <td>0.304706</td>\n",
       "      <td>-0.565719</td>\n",
       "      <td>-1.458590</td>\n",
       "      <td>-0.413996</td>\n",
       "      <td>-0.617453</td>\n",
       "      <td>0.188299</td>\n",
       "      <td>-0.540118</td>\n",
       "      <td>-1.077096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.498299</td>\n",
       "      <td>0.769775</td>\n",
       "      <td>0.871095</td>\n",
       "      <td>-1.349253</td>\n",
       "      <td>0.890614</td>\n",
       "      <td>0.869422</td>\n",
       "      <td>-0.422121</td>\n",
       "      <td>0.827618</td>\n",
       "      <td>-1.231198</td>\n",
       "      <td>-0.852264</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.562223</td>\n",
       "      <td>-1.791608</td>\n",
       "      <td>0.672705</td>\n",
       "      <td>-0.442705</td>\n",
       "      <td>0.240579</td>\n",
       "      <td>-1.214454</td>\n",
       "      <td>-1.424031</td>\n",
       "      <td>-0.364616</td>\n",
       "      <td>-0.704214</td>\n",
       "      <td>-1.668375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.267042</td>\n",
       "      <td>0.421862</td>\n",
       "      <td>-0.119198</td>\n",
       "      <td>-0.072423</td>\n",
       "      <td>0.397617</td>\n",
       "      <td>1.587639</td>\n",
       "      <td>-0.025031</td>\n",
       "      <td>1.552775</td>\n",
       "      <td>-1.048442</td>\n",
       "      <td>0.563937</td>\n",
       "      <td>...</td>\n",
       "      <td>0.409026</td>\n",
       "      <td>-0.570773</td>\n",
       "      <td>-0.076436</td>\n",
       "      <td>0.282434</td>\n",
       "      <td>-0.453448</td>\n",
       "      <td>-0.363573</td>\n",
       "      <td>-0.163451</td>\n",
       "      <td>0.155775</td>\n",
       "      <td>-0.598525</td>\n",
       "      <td>0.598194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.072785</td>\n",
       "      <td>0.047187</td>\n",
       "      <td>-0.868375</td>\n",
       "      <td>0.531483</td>\n",
       "      <td>1.365352</td>\n",
       "      <td>0.308314</td>\n",
       "      <td>-0.948498</td>\n",
       "      <td>-0.299164</td>\n",
       "      <td>-1.657629</td>\n",
       "      <td>0.004155</td>\n",
       "      <td>...</td>\n",
       "      <td>1.054163</td>\n",
       "      <td>-1.057066</td>\n",
       "      <td>0.567562</td>\n",
       "      <td>-0.367169</td>\n",
       "      <td>0.384171</td>\n",
       "      <td>-0.920323</td>\n",
       "      <td>-0.322835</td>\n",
       "      <td>-0.364616</td>\n",
       "      <td>-0.474758</td>\n",
       "      <td>0.236857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.128287</td>\n",
       "      <td>1.050781</td>\n",
       "      <td>-0.937265</td>\n",
       "      <td>-0.279477</td>\n",
       "      <td>0.653245</td>\n",
       "      <td>1.161198</td>\n",
       "      <td>-0.565259</td>\n",
       "      <td>0.180555</td>\n",
       "      <td>-1.194647</td>\n",
       "      <td>-0.646532</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.702042</td>\n",
       "      <td>-0.098082</td>\n",
       "      <td>0.146992</td>\n",
       "      <td>-0.144880</td>\n",
       "      <td>0.360239</td>\n",
       "      <td>-1.416145</td>\n",
       "      <td>-1.337095</td>\n",
       "      <td>0.383446</td>\n",
       "      <td>-0.540118</td>\n",
       "      <td>0.368253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.035784</td>\n",
       "      <td>-0.916264</td>\n",
       "      <td>-0.282811</td>\n",
       "      <td>0.652264</td>\n",
       "      <td>0.534561</td>\n",
       "      <td>0.278388</td>\n",
       "      <td>1.253971</td>\n",
       "      <td>0.604493</td>\n",
       "      <td>-0.414888</td>\n",
       "      <td>-0.215930</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.191867</td>\n",
       "      <td>-0.407542</td>\n",
       "      <td>-0.457578</td>\n",
       "      <td>0.120572</td>\n",
       "      <td>-1.003883</td>\n",
       "      <td>-0.279536</td>\n",
       "      <td>0.527212</td>\n",
       "      <td>0.741215</td>\n",
       "      <td>-0.359335</td>\n",
       "      <td>-0.452968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>-3.285076</td>\n",
       "      <td>-0.835977</td>\n",
       "      <td>-0.007252</td>\n",
       "      <td>1.031862</td>\n",
       "      <td>-1.154411</td>\n",
       "      <td>0.338240</td>\n",
       "      <td>0.007291</td>\n",
       "      <td>0.637962</td>\n",
       "      <td>-0.012825</td>\n",
       "      <td>0.291223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.743542</td>\n",
       "      <td>1.578440</td>\n",
       "      <td>1.553274</td>\n",
       "      <td>-0.311057</td>\n",
       "      <td>1.676497</td>\n",
       "      <td>1.747872</td>\n",
       "      <td>3.734206</td>\n",
       "      <td>1.359179</td>\n",
       "      <td>0.030045</td>\n",
       "      <td>-1.766921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.368795</td>\n",
       "      <td>0.007043</td>\n",
       "      <td>-0.876987</td>\n",
       "      <td>-0.892010</td>\n",
       "      <td>1.137113</td>\n",
       "      <td>0.742237</td>\n",
       "      <td>0.034995</td>\n",
       "      <td>0.671431</td>\n",
       "      <td>-0.037192</td>\n",
       "      <td>-0.799634</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086458</td>\n",
       "      <td>-0.604779</td>\n",
       "      <td>-0.865005</td>\n",
       "      <td>0.636370</td>\n",
       "      <td>0.025192</td>\n",
       "      <td>-0.489630</td>\n",
       "      <td>-0.631943</td>\n",
       "      <td>-0.071896</td>\n",
       "      <td>-0.524821</td>\n",
       "      <td>0.532497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.711056</td>\n",
       "      <td>1.331788</td>\n",
       "      <td>1.008875</td>\n",
       "      <td>0.220902</td>\n",
       "      <td>1.191890</td>\n",
       "      <td>1.632528</td>\n",
       "      <td>0.584457</td>\n",
       "      <td>0.035524</td>\n",
       "      <td>-0.792584</td>\n",
       "      <td>-0.632178</td>\n",
       "      <td>...</td>\n",
       "      <td>1.663458</td>\n",
       "      <td>-1.203295</td>\n",
       "      <td>-0.746720</td>\n",
       "      <td>3.323266</td>\n",
       "      <td>0.049123</td>\n",
       "      <td>-0.695522</td>\n",
       "      <td>-1.187371</td>\n",
       "      <td>-2.251033</td>\n",
       "      <td>-0.238349</td>\n",
       "      <td>0.171160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1.275325</td>\n",
       "      <td>-1.050077</td>\n",
       "      <td>0.828039</td>\n",
       "      <td>0.315802</td>\n",
       "      <td>0.132859</td>\n",
       "      <td>0.263425</td>\n",
       "      <td>1.706469</td>\n",
       "      <td>0.002055</td>\n",
       "      <td>-1.133728</td>\n",
       "      <td>1.707423</td>\n",
       "      <td>...</td>\n",
       "      <td>0.552390</td>\n",
       "      <td>0.915313</td>\n",
       "      <td>-1.561575</td>\n",
       "      <td>-0.410332</td>\n",
       "      <td>-0.740631</td>\n",
       "      <td>1.189022</td>\n",
       "      <td>-0.390452</td>\n",
       "      <td>1.098983</td>\n",
       "      <td>0.056467</td>\n",
       "      <td>0.433950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>-0.213974</td>\n",
       "      <td>-0.648639</td>\n",
       "      <td>-1.204214</td>\n",
       "      <td>-0.426140</td>\n",
       "      <td>1.146242</td>\n",
       "      <td>0.689867</td>\n",
       "      <td>-0.343627</td>\n",
       "      <td>1.976713</td>\n",
       "      <td>0.328320</td>\n",
       "      <td>-1.498166</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.905140</td>\n",
       "      <td>-0.529965</td>\n",
       "      <td>-0.273578</td>\n",
       "      <td>-0.783692</td>\n",
       "      <td>1.102130</td>\n",
       "      <td>-1.388833</td>\n",
       "      <td>-1.076285</td>\n",
       "      <td>-0.885007</td>\n",
       "      <td>-0.670839</td>\n",
       "      <td>0.236857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-1.166756</td>\n",
       "      <td>0.809919</td>\n",
       "      <td>-1.298937</td>\n",
       "      <td>-1.771988</td>\n",
       "      <td>-0.287101</td>\n",
       "      <td>-0.985974</td>\n",
       "      <td>-0.583728</td>\n",
       "      <td>-0.254539</td>\n",
       "      <td>0.303952</td>\n",
       "      <td>-1.096271</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.248057</td>\n",
       "      <td>-1.186291</td>\n",
       "      <td>0.633277</td>\n",
       "      <td>-0.315374</td>\n",
       "      <td>-0.764563</td>\n",
       "      <td>0.296122</td>\n",
       "      <td>0.334019</td>\n",
       "      <td>-0.006847</td>\n",
       "      <td>-0.325959</td>\n",
       "      <td>0.072613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.433547</td>\n",
       "      <td>-2.281153</td>\n",
       "      <td>0.053027</td>\n",
       "      <td>-1.875514</td>\n",
       "      <td>-0.369267</td>\n",
       "      <td>-0.447311</td>\n",
       "      <td>-0.435973</td>\n",
       "      <td>0.916868</td>\n",
       "      <td>0.194298</td>\n",
       "      <td>-0.617825</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.594519</td>\n",
       "      <td>1.044538</td>\n",
       "      <td>1.724131</td>\n",
       "      <td>-0.481551</td>\n",
       "      <td>-0.046604</td>\n",
       "      <td>-0.529548</td>\n",
       "      <td>0.000763</td>\n",
       "      <td>-0.332092</td>\n",
       "      <td>-0.363507</td>\n",
       "      <td>2.142089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.008033</td>\n",
       "      <td>-1.505040</td>\n",
       "      <td>-0.833930</td>\n",
       "      <td>0.635009</td>\n",
       "      <td>-0.597507</td>\n",
       "      <td>0.323277</td>\n",
       "      <td>-1.701123</td>\n",
       "      <td>0.537556</td>\n",
       "      <td>1.034976</td>\n",
       "      <td>-1.311572</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.033011</td>\n",
       "      <td>-0.087880</td>\n",
       "      <td>0.054992</td>\n",
       "      <td>-0.324006</td>\n",
       "      <td>0.503831</td>\n",
       "      <td>-0.468620</td>\n",
       "      <td>0.266402</td>\n",
       "      <td>2.009667</td>\n",
       "      <td>-0.419133</td>\n",
       "      <td>-2.128258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.368795</td>\n",
       "      <td>0.007043</td>\n",
       "      <td>-0.584205</td>\n",
       "      <td>1.161270</td>\n",
       "      <td>-1.291355</td>\n",
       "      <td>-0.163017</td>\n",
       "      <td>-0.103525</td>\n",
       "      <td>0.191712</td>\n",
       "      <td>0.791301</td>\n",
       "      <td>-0.560411</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.486996</td>\n",
       "      <td>-1.220298</td>\n",
       "      <td>-0.010722</td>\n",
       "      <td>0.230638</td>\n",
       "      <td>1.054266</td>\n",
       "      <td>0.346545</td>\n",
       "      <td>-1.288797</td>\n",
       "      <td>0.155775</td>\n",
       "      <td>0.420816</td>\n",
       "      <td>-0.387270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.082036</td>\n",
       "      <td>-1.103602</td>\n",
       "      <td>-0.523926</td>\n",
       "      <td>-1.099064</td>\n",
       "      <td>0.297192</td>\n",
       "      <td>1.228531</td>\n",
       "      <td>-2.010484</td>\n",
       "      <td>1.039587</td>\n",
       "      <td>0.450157</td>\n",
       "      <td>-0.947953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.134246</td>\n",
       "      <td>-0.247711</td>\n",
       "      <td>0.896133</td>\n",
       "      <td>4.592258</td>\n",
       "      <td>-0.381652</td>\n",
       "      <td>-0.065240</td>\n",
       "      <td>-0.110323</td>\n",
       "      <td>0.058202</td>\n",
       "      <td>-0.499790</td>\n",
       "      <td>0.335404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>-1.407264</td>\n",
       "      <td>-2.361441</td>\n",
       "      <td>-0.506703</td>\n",
       "      <td>-0.866128</td>\n",
       "      <td>-1.483076</td>\n",
       "      <td>-1.322639</td>\n",
       "      <td>2.135881</td>\n",
       "      <td>-0.555758</td>\n",
       "      <td>0.803485</td>\n",
       "      <td>1.138072</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.702042</td>\n",
       "      <td>0.221579</td>\n",
       "      <td>0.475563</td>\n",
       "      <td>-0.727580</td>\n",
       "      <td>-1.219270</td>\n",
       "      <td>-0.600980</td>\n",
       "      <td>1.647728</td>\n",
       "      <td>-0.657336</td>\n",
       "      <td>-0.412179</td>\n",
       "      <td>1.288020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1.081069</td>\n",
       "      <td>0.462006</td>\n",
       "      <td>-0.928654</td>\n",
       "      <td>0.514228</td>\n",
       "      <td>0.781059</td>\n",
       "      <td>0.211055</td>\n",
       "      <td>0.058081</td>\n",
       "      <td>-0.433039</td>\n",
       "      <td>0.620729</td>\n",
       "      <td>-0.421662</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.068852</td>\n",
       "      <td>0.014140</td>\n",
       "      <td>-0.418149</td>\n",
       "      <td>0.375234</td>\n",
       "      <td>1.485041</td>\n",
       "      <td>0.247800</td>\n",
       "      <td>-0.274537</td>\n",
       "      <td>0.090726</td>\n",
       "      <td>0.219173</td>\n",
       "      <td>-0.551514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1.256825</td>\n",
       "      <td>0.488768</td>\n",
       "      <td>0.164973</td>\n",
       "      <td>0.100121</td>\n",
       "      <td>2.059200</td>\n",
       "      <td>2.126303</td>\n",
       "      <td>0.335121</td>\n",
       "      <td>1.073056</td>\n",
       "      <td>-0.695114</td>\n",
       "      <td>1.970568</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.068852</td>\n",
       "      <td>-0.880232</td>\n",
       "      <td>0.593848</td>\n",
       "      <td>-0.263578</td>\n",
       "      <td>0.168783</td>\n",
       "      <td>-1.239666</td>\n",
       "      <td>-1.375733</td>\n",
       "      <td>-0.429665</td>\n",
       "      <td>2.752925</td>\n",
       "      <td>0.893834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>-1.823528</td>\n",
       "      <td>-0.554970</td>\n",
       "      <td>-0.670317</td>\n",
       "      <td>1.126762</td>\n",
       "      <td>1.255797</td>\n",
       "      <td>0.712311</td>\n",
       "      <td>1.697235</td>\n",
       "      <td>-0.399570</td>\n",
       "      <td>0.072461</td>\n",
       "      <td>0.262516</td>\n",
       "      <td>...</td>\n",
       "      <td>0.552390</td>\n",
       "      <td>-0.047072</td>\n",
       "      <td>0.712133</td>\n",
       "      <td>0.157261</td>\n",
       "      <td>1.269653</td>\n",
       "      <td>-0.277435</td>\n",
       "      <td>-0.525687</td>\n",
       "      <td>-0.917531</td>\n",
       "      <td>-0.185504</td>\n",
       "      <td>-0.058782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0.822060</td>\n",
       "      <td>1.358550</td>\n",
       "      <td>0.569701</td>\n",
       "      <td>0.419329</td>\n",
       "      <td>-0.825746</td>\n",
       "      <td>-0.514644</td>\n",
       "      <td>-1.410231</td>\n",
       "      <td>0.080149</td>\n",
       "      <td>-0.792584</td>\n",
       "      <td>-0.804419</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.705587</td>\n",
       "      <td>-0.540167</td>\n",
       "      <td>0.633277</td>\n",
       "      <td>-0.397383</td>\n",
       "      <td>-0.046604</td>\n",
       "      <td>-0.590475</td>\n",
       "      <td>0.121508</td>\n",
       "      <td>-1.470446</td>\n",
       "      <td>-0.516478</td>\n",
       "      <td>-0.091631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>-0.028968</td>\n",
       "      <td>0.957113</td>\n",
       "      <td>-0.170865</td>\n",
       "      <td>1.282052</td>\n",
       "      <td>0.808448</td>\n",
       "      <td>0.121278</td>\n",
       "      <td>0.058081</td>\n",
       "      <td>-0.243383</td>\n",
       "      <td>-2.205897</td>\n",
       "      <td>0.305576</td>\n",
       "      <td>...</td>\n",
       "      <td>2.021867</td>\n",
       "      <td>0.636459</td>\n",
       "      <td>-0.313007</td>\n",
       "      <td>0.073093</td>\n",
       "      <td>0.432035</td>\n",
       "      <td>0.632273</td>\n",
       "      <td>-0.250388</td>\n",
       "      <td>0.643641</td>\n",
       "      <td>-0.287021</td>\n",
       "      <td>0.860985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>-0.130721</td>\n",
       "      <td>-0.474683</td>\n",
       "      <td>-0.455036</td>\n",
       "      <td>1.342442</td>\n",
       "      <td>-1.190929</td>\n",
       "      <td>-0.948567</td>\n",
       "      <td>-0.449825</td>\n",
       "      <td>-1.269759</td>\n",
       "      <td>1.303018</td>\n",
       "      <td>0.999323</td>\n",
       "      <td>...</td>\n",
       "      <td>0.361239</td>\n",
       "      <td>-0.397340</td>\n",
       "      <td>-0.878148</td>\n",
       "      <td>-0.052080</td>\n",
       "      <td>-0.070536</td>\n",
       "      <td>3.037850</td>\n",
       "      <td>-0.414601</td>\n",
       "      <td>-1.015104</td>\n",
       "      <td>0.397175</td>\n",
       "      <td>0.631043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>0.933064</td>\n",
       "      <td>-0.755689</td>\n",
       "      <td>-0.282811</td>\n",
       "      <td>-0.693584</td>\n",
       "      <td>-0.844005</td>\n",
       "      <td>-0.671754</td>\n",
       "      <td>0.265861</td>\n",
       "      <td>-1.827572</td>\n",
       "      <td>0.657280</td>\n",
       "      <td>-0.397740</td>\n",
       "      <td>...</td>\n",
       "      <td>0.970534</td>\n",
       "      <td>0.877906</td>\n",
       "      <td>-0.838720</td>\n",
       "      <td>-0.561403</td>\n",
       "      <td>-0.884223</td>\n",
       "      <td>-0.073644</td>\n",
       "      <td>1.068151</td>\n",
       "      <td>-1.015104</td>\n",
       "      <td>-0.235568</td>\n",
       "      <td>-1.175642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1.182822</td>\n",
       "      <td>0.676106</td>\n",
       "      <td>-0.343090</td>\n",
       "      <td>-0.426140</td>\n",
       "      <td>-0.917042</td>\n",
       "      <td>-0.290201</td>\n",
       "      <td>-1.419466</td>\n",
       "      <td>-0.990852</td>\n",
       "      <td>2.241166</td>\n",
       "      <td>0.645273</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.024610</td>\n",
       "      <td>-1.410734</td>\n",
       "      <td>-0.707291</td>\n",
       "      <td>-0.494500</td>\n",
       "      <td>0.767082</td>\n",
       "      <td>-0.245921</td>\n",
       "      <td>-0.197260</td>\n",
       "      <td>0.285873</td>\n",
       "      <td>1.465189</td>\n",
       "      <td>-1.241340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>-0.852245</td>\n",
       "      <td>-1.357846</td>\n",
       "      <td>1.318879</td>\n",
       "      <td>1.178525</td>\n",
       "      <td>0.616727</td>\n",
       "      <td>-0.881234</td>\n",
       "      <td>0.385912</td>\n",
       "      <td>-0.466508</td>\n",
       "      <td>0.243033</td>\n",
       "      <td>-1.163254</td>\n",
       "      <td>...</td>\n",
       "      <td>0.659913</td>\n",
       "      <td>3.067927</td>\n",
       "      <td>0.764705</td>\n",
       "      <td>0.187475</td>\n",
       "      <td>-1.626114</td>\n",
       "      <td>0.890688</td>\n",
       "      <td>0.029741</td>\n",
       "      <td>0.220824</td>\n",
       "      <td>-0.359335</td>\n",
       "      <td>1.715054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0.886812</td>\n",
       "      <td>1.746607</td>\n",
       "      <td>1.215544</td>\n",
       "      <td>1.238916</td>\n",
       "      <td>-1.090504</td>\n",
       "      <td>-0.185461</td>\n",
       "      <td>-0.819212</td>\n",
       "      <td>-2.106478</td>\n",
       "      <td>-1.219015</td>\n",
       "      <td>1.176348</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.391420</td>\n",
       "      <td>-0.900636</td>\n",
       "      <td>-0.536435</td>\n",
       "      <td>-0.140564</td>\n",
       "      <td>1.317517</td>\n",
       "      <td>1.815102</td>\n",
       "      <td>-0.153791</td>\n",
       "      <td>-0.234518</td>\n",
       "      <td>3.738891</td>\n",
       "      <td>-0.124480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>0.452048</td>\n",
       "      <td>0.154237</td>\n",
       "      <td>0.139139</td>\n",
       "      <td>1.816940</td>\n",
       "      <td>-0.433174</td>\n",
       "      <td>-0.529607</td>\n",
       "      <td>-1.414848</td>\n",
       "      <td>-0.009101</td>\n",
       "      <td>1.826919</td>\n",
       "      <td>2.147593</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.905140</td>\n",
       "      <td>1.258779</td>\n",
       "      <td>1.224704</td>\n",
       "      <td>-0.149196</td>\n",
       "      <td>-2.200481</td>\n",
       "      <td>1.466346</td>\n",
       "      <td>1.406237</td>\n",
       "      <td>0.058202</td>\n",
       "      <td>2.802988</td>\n",
       "      <td>0.959531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38 rows Ã 7129 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0         1         2         3         4         5         6     \\\n",
       "0  -0.861496 -0.033101 -0.351701 -0.805738 -0.168417 -0.888716  1.120068   \n",
       "1  -0.167723  1.037400  0.139139  0.876572  0.114600  0.293351 -1.322502   \n",
       "2   0.415047  1.358550 -2.495899  1.100880 -0.907912 -1.577008  0.353591   \n",
       "3  -0.130721  0.488768  2.429729 -1.461407 -1.300484 -1.090715  0.930757   \n",
       "4   0.137537  0.341574 -0.506703 -0.115559  0.425006  1.161198  0.219688   \n",
       "5  -0.158472  0.876825  1.999167 -0.952401  0.041564 -0.888716  0.510580   \n",
       "6   0.452048  0.087330  2.197225 -1.090436 -1.117893 -0.836346  0.806089   \n",
       "7  -2.702307 -1.464896  0.208029 -1.582188 -2.414292 -2.624409 -1.068548   \n",
       "8   1.164322  0.314812  1.060542  0.747163  0.607597 -0.716643  0.201219   \n",
       "9   0.304043  0.609200  0.509423  0.324429  0.899744  1.445492  1.715704   \n",
       "10 -0.408231 -0.059864 -0.463647 -0.857501  1.036687  0.068908  0.662952   \n",
       "11  0.498299  0.769775  0.871095 -1.349253  0.890614  0.869422 -0.422121   \n",
       "12  0.267042  0.421862 -0.119198 -0.072423  0.397617  1.587639 -0.025031   \n",
       "13  0.072785  0.047187 -0.868375  0.531483  1.365352  0.308314 -0.948498   \n",
       "14  0.128287  1.050781 -0.937265 -0.279477  0.653245  1.161198 -0.565259   \n",
       "15  0.035784 -0.916264 -0.282811  0.652264  0.534561  0.278388  1.253971   \n",
       "16 -3.285076 -0.835977 -0.007252  1.031862 -1.154411  0.338240  0.007291   \n",
       "17  0.368795  0.007043 -0.876987 -0.892010  1.137113  0.742237  0.034995   \n",
       "18  0.711056  1.331788  1.008875  0.220902  1.191890  1.632528  0.584457   \n",
       "19  1.275325 -1.050077  0.828039  0.315802  0.132859  0.263425  1.706469   \n",
       "20 -0.213974 -0.648639 -1.204214 -0.426140  1.146242  0.689867 -0.343627   \n",
       "21 -1.166756  0.809919 -1.298937 -1.771988 -0.287101 -0.985974 -0.583728   \n",
       "22  0.433547 -2.281153  0.053027 -1.875514 -0.369267 -0.447311 -0.435973   \n",
       "23  0.008033 -1.505040 -0.833930  0.635009 -0.597507  0.323277 -1.701123   \n",
       "24  0.368795  0.007043 -0.584205  1.161270 -1.291355 -0.163017 -0.103525   \n",
       "25  0.082036 -1.103602 -0.523926 -1.099064  0.297192  1.228531 -2.010484   \n",
       "26 -1.407264 -2.361441 -0.506703 -0.866128 -1.483076 -1.322639  2.135881   \n",
       "27  1.081069  0.462006 -0.928654  0.514228  0.781059  0.211055  0.058081   \n",
       "28  1.256825  0.488768  0.164973  0.100121  2.059200  2.126303  0.335121   \n",
       "29 -1.823528 -0.554970 -0.670317  1.126762  1.255797  0.712311  1.697235   \n",
       "30  0.822060  1.358550  0.569701  0.419329 -0.825746 -0.514644 -1.410231   \n",
       "31 -0.028968  0.957113 -0.170865  1.282052  0.808448  0.121278  0.058081   \n",
       "32 -0.130721 -0.474683 -0.455036  1.342442 -1.190929 -0.948567 -0.449825   \n",
       "33  0.933064 -0.755689 -0.282811 -0.693584 -0.844005 -0.671754  0.265861   \n",
       "34  1.182822  0.676106 -0.343090 -0.426140 -0.917042 -0.290201 -1.419466   \n",
       "35 -0.852245 -1.357846  1.318879  1.178525  0.616727 -0.881234  0.385912   \n",
       "36  0.886812  1.746607  1.215544  1.238916 -1.090504 -0.185461 -0.819212   \n",
       "37  0.452048  0.154237  0.139139  1.816940 -0.433174 -0.529607 -1.414848   \n",
       "\n",
       "        7         8         9     ...      7119      7120      7121      7122  \\\n",
       "0   0.280962  1.863470  0.449110  ...  0.074511 -0.815620 -1.758717 -0.021865   \n",
       "1   0.370212  0.023726 -0.182439  ... -0.116640  0.292993 -0.589006  0.092516   \n",
       "2  -1.849884  1.303018 -1.565148  ...  1.627617  1.524030  0.317849 -0.498816   \n",
       "3  -0.578070 -0.609828 -0.388171  ...  0.731595  0.286192  2.749271 -0.485868   \n",
       "4   0.883400 -0.353970  0.669195  ... -0.271951 -0.346330  0.633277  0.226322   \n",
       "5   0.169399 -0.146846  0.386912  ... -0.761776  1.598844 -1.114719 -0.490184   \n",
       "6   0.247493  0.328320 -0.632178  ... -1.777269  0.231781 -2.455287 -0.535505   \n",
       "7  -2.920885 -0.353970 -1.345063  ...  1.316996 -0.414343 -1.246147 -0.209624   \n",
       "8   0.303274 -0.914421  1.884448  ...  2.117443  0.779287 -0.037007 -0.479393   \n",
       "9   0.593337  0.949690  0.339067  ... -1.633905  0.799691  0.225849 -0.643412   \n",
       "10  1.028431 -0.524542  1.138072  ...  0.074511 -0.370134  0.304706 -0.565719   \n",
       "11  0.827618 -1.231198 -0.852264  ... -1.562223 -1.791608  0.672705 -0.442705   \n",
       "12  1.552775 -1.048442  0.563937  ...  0.409026 -0.570773 -0.076436  0.282434   \n",
       "13 -0.299164 -1.657629  0.004155  ...  1.054163 -1.057066  0.567562 -0.367169   \n",
       "14  0.180555 -1.194647 -0.646532  ... -0.702042 -0.098082  0.146992 -0.144880   \n",
       "15  0.604493 -0.414888 -0.215930  ... -1.191867 -0.407542 -0.457578  0.120572   \n",
       "16  0.637962 -0.012825  0.291223  ...  0.743542  1.578440  1.553274 -0.311057   \n",
       "17  0.671431 -0.037192 -0.799634  ...  0.086458 -0.604779 -0.865005  0.636370   \n",
       "18  0.035524 -0.792584 -0.632178  ...  1.663458 -1.203295 -0.746720  3.323266   \n",
       "19  0.002055 -1.133728  1.707423  ...  0.552390  0.915313 -1.561575 -0.410332   \n",
       "20  1.976713  0.328320 -1.498166  ... -0.905140 -0.529965 -0.273578 -0.783692   \n",
       "21 -0.254539  0.303952 -1.096271  ... -0.248057 -1.186291  0.633277 -0.315374   \n",
       "22  0.916868  0.194298 -0.617825  ... -0.594519  1.044538  1.724131 -0.481551   \n",
       "23  0.537556  1.034976 -1.311572  ... -0.033011 -0.087880  0.054992 -0.324006   \n",
       "24  0.191712  0.791301 -0.560411  ... -0.486996 -1.220298 -0.010722  0.230638   \n",
       "25  1.039587  0.450157 -0.947953  ...  0.134246 -0.247711  0.896133  4.592258   \n",
       "26 -0.555758  0.803485  1.138072  ... -0.702042  0.221579  0.475563 -0.727580   \n",
       "27 -0.433039  0.620729 -0.421662  ... -0.068852  0.014140 -0.418149  0.375234   \n",
       "28  1.073056 -0.695114  1.970568  ... -0.068852 -0.880232  0.593848 -0.263578   \n",
       "29 -0.399570  0.072461  0.262516  ...  0.552390 -0.047072  0.712133  0.157261   \n",
       "30  0.080149 -0.792584 -0.804419  ... -1.705587 -0.540167  0.633277 -0.397383   \n",
       "31 -0.243383 -2.205897  0.305576  ...  2.021867  0.636459 -0.313007  0.073093   \n",
       "32 -1.269759  1.303018  0.999323  ...  0.361239 -0.397340 -0.878148 -0.052080   \n",
       "33 -1.827572  0.657280 -0.397740  ...  0.970534  0.877906 -0.838720 -0.561403   \n",
       "34 -0.990852  2.241166  0.645273  ... -1.024610 -1.410734 -0.707291 -0.494500   \n",
       "35 -0.466508  0.243033 -1.163254  ...  0.659913  3.067927  0.764705  0.187475   \n",
       "36 -2.106478 -1.219015  1.176348  ... -0.391420 -0.900636 -0.536435 -0.140564   \n",
       "37 -0.009101  1.826919  2.147593  ... -0.905140  1.258779  1.224704 -0.149196   \n",
       "\n",
       "        7123      7124      7125      7126      7127      7128  \n",
       "0  -0.405584 -0.159782 -0.033046  0.546068 -0.435820 -0.255875  \n",
       "1   0.073055 -0.182892 -0.197260 -0.267043 -0.595744  0.499648  \n",
       "2   1.724361  0.565043  2.130709  0.708690 -0.384366 -0.387270  \n",
       "3  -2.152617 -0.508538 -0.800986 -2.251033 -0.526212 -2.029712  \n",
       "4  -0.142332 -1.300593 -0.105493 -0.169469 -0.623557  0.138311  \n",
       "5  -1.291066 -0.470721  0.024912  0.220824 -0.433039 -0.781456  \n",
       "6   0.049123  0.569244  0.705915 -0.299567 -0.188286 -0.420119  \n",
       "7   1.030334  1.953764  0.532042  1.294130  0.384659  1.616508  \n",
       "8   0.192715 -0.233315  0.237423  3.115498 -0.362116 -0.321573  \n",
       "9   0.982470 -0.628292  0.111848 -0.332092 -0.463633  1.189473  \n",
       "10 -1.458590 -0.413996 -0.617453  0.188299 -0.540118 -1.077096  \n",
       "11  0.240579 -1.214454 -1.424031 -0.364616 -0.704214 -1.668375  \n",
       "12 -0.453448 -0.363573 -0.163451  0.155775 -0.598525  0.598194  \n",
       "13  0.384171 -0.920323 -0.322835 -0.364616 -0.474758  0.236857  \n",
       "14  0.360239 -1.416145 -1.337095  0.383446 -0.540118  0.368253  \n",
       "15 -1.003883 -0.279536  0.527212  0.741215 -0.359335 -0.452968  \n",
       "16  1.676497  1.747872  3.734206  1.359179  0.030045 -1.766921  \n",
       "17  0.025192 -0.489630 -0.631943 -0.071896 -0.524821  0.532497  \n",
       "18  0.049123 -0.695522 -1.187371 -2.251033 -0.238349  0.171160  \n",
       "19 -0.740631  1.189022 -0.390452  1.098983  0.056467  0.433950  \n",
       "20  1.102130 -1.388833 -1.076285 -0.885007 -0.670839  0.236857  \n",
       "21 -0.764563  0.296122  0.334019 -0.006847 -0.325959  0.072613  \n",
       "22 -0.046604 -0.529548  0.000763 -0.332092 -0.363507  2.142089  \n",
       "23  0.503831 -0.468620  0.266402  2.009667 -0.419133 -2.128258  \n",
       "24  1.054266  0.346545 -1.288797  0.155775  0.420816 -0.387270  \n",
       "25 -0.381652 -0.065240 -0.110323  0.058202 -0.499790  0.335404  \n",
       "26 -1.219270 -0.600980  1.647728 -0.657336 -0.412179  1.288020  \n",
       "27  1.485041  0.247800 -0.274537  0.090726  0.219173 -0.551514  \n",
       "28  0.168783 -1.239666 -1.375733 -0.429665  2.752925  0.893834  \n",
       "29  1.269653 -0.277435 -0.525687 -0.917531 -0.185504 -0.058782  \n",
       "30 -0.046604 -0.590475  0.121508 -1.470446 -0.516478 -0.091631  \n",
       "31  0.432035  0.632273 -0.250388  0.643641 -0.287021  0.860985  \n",
       "32 -0.070536  3.037850 -0.414601 -1.015104  0.397175  0.631043  \n",
       "33 -0.884223 -0.073644  1.068151 -1.015104 -0.235568 -1.175642  \n",
       "34  0.767082 -0.245921 -0.197260  0.285873  1.465189 -1.241340  \n",
       "35 -1.626114  0.890688  0.029741  0.220824 -0.359335  1.715054  \n",
       "36  1.317517  1.815102 -0.153791 -0.234518  3.738891 -0.124480  \n",
       "37 -2.200481  1.466346  1.406237  0.058202  2.802988  0.959531  \n",
       "\n",
       "[38 rows x 7129 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_cudf_train"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Model Building\n",
    "#### Dask Integration\n",
    "\n",
    "We will try using the Random Forests Classifier  and implement using CuML and Dask."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Start Dask cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "NVMLError_NotFound",
     "evalue": "Not Found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNVMLError_NotFound\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-a08cbce908ea>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# This will use all GPUs on the local host by default\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mcluster\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mLocalCUDACluster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mthreads_per_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#Set 1 thread per worker using arguments to cluster\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#Pass the cluster as an argument to Client\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/local_cuda_cluster.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, CUDA_VISIBLE_DEVICES, n_workers, threads_per_worker, memory_limit, device_memory_limit, data, local_directory, shared_filesystem, protocol, enable_tcp_over_ucx, enable_infiniband, enable_nvlink, enable_rdmacm, ucx_net_devices, rmm_pool_size, rmm_managed_memory, rmm_async, rmm_log_directory, jit_unspill, log_spilling, worker_class, **kwargs)\u001b[0m\n\u001b[1;32m    225\u001b[0m         )\n\u001b[1;32m    226\u001b[0m         self.device_memory_limit = parse_device_memory_limit(\n\u001b[0;32m--> 227\u001b[0;31m             \u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdevice_index\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnvml_device_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCUDA_VISIBLE_DEVICES\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    228\u001b[0m         )\n\u001b[1;32m    229\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/utils.py\u001b[0m in \u001b[0;36mparse_device_memory_limit\u001b[0;34m(device_memory_limit, device_index)\u001b[0m\n\u001b[1;32m    624\u001b[0m         \u001b[0mdevice_memory_limit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    625\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mdevice_memory_limit\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 626\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_device_total_memory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_index\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    627\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    628\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/utils.py\u001b[0m in \u001b[0;36mget_device_total_memory\u001b[0;34m(index)\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misnumeric\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    235\u001b[0m         \u001b[0;31m# This means index is UUID. This works for both MIG and non-MIG device UUIDs.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 236\u001b[0;31m         \u001b[0mhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpynvml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnvmlDeviceGetHandleByUUID\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    237\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    238\u001b[0m         \u001b[0;31m# This is a device index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/pynvml/nvml.py\u001b[0m in \u001b[0;36mnvmlDeviceGetHandleByUUID\u001b[0;34m(uuid)\u001b[0m\n\u001b[1;32m   1669\u001b[0m     \u001b[0mfn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_nvmlGetFunctionPointer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"nvmlDeviceGetHandleByUUID\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1670\u001b[0m     \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc_uuid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbyref\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1671\u001b[0;31m     \u001b[0m_nvmlCheckReturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1672\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdevice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1673\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/pynvml/nvml.py\u001b[0m in \u001b[0;36m_nvmlCheckReturn\u001b[0;34m(ret)\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_nvmlCheckReturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mNVML_SUCCESS\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 765\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mNVMLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    766\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNVMLError_NotFound\u001b[0m: Not Found"
     ]
    }
   ],
   "source": [
    "#Modify the code in this cell\n",
    "\n",
    "# This will use all GPUs on the local host by default\n",
    "cluster = LocalCUDACluster(threads_per_worker=1) #Set 1 thread per worker using arguments to cluster\n",
    "c = Client(cluster) #Pass the cluster as an argument to Client\n",
    "\n",
    "# Query the client for all connected workers\n",
    "workers = c.has_what().keys()\n",
    "n_workers = len(workers)\n",
    "n_streams = 8 # Performance optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "NVMLError_NotFound",
     "evalue": "Not Found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNVMLError_NotFound\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-26-58602d71ddce>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# This will use all GPUs on the local host by default\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mcluster\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mLocalCUDACluster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mthreads_per_worker\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClient\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcluster\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m# Query the client for all connected workers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/local_cuda_cluster.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, CUDA_VISIBLE_DEVICES, n_workers, threads_per_worker, memory_limit, device_memory_limit, data, local_directory, shared_filesystem, protocol, enable_tcp_over_ucx, enable_infiniband, enable_nvlink, enable_rdmacm, ucx_net_devices, rmm_pool_size, rmm_managed_memory, rmm_async, rmm_log_directory, jit_unspill, log_spilling, worker_class, **kwargs)\u001b[0m\n\u001b[1;32m    225\u001b[0m         )\n\u001b[1;32m    226\u001b[0m         self.device_memory_limit = parse_device_memory_limit(\n\u001b[0;32m--> 227\u001b[0;31m             \u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdevice_index\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnvml_device_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mCUDA_VISIBLE_DEVICES\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    228\u001b[0m         )\n\u001b[1;32m    229\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/utils.py\u001b[0m in \u001b[0;36mparse_device_memory_limit\u001b[0;34m(device_memory_limit, device_index)\u001b[0m\n\u001b[1;32m    624\u001b[0m         \u001b[0mdevice_memory_limit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    625\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfloat\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mdevice_memory_limit\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 626\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_device_total_memory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_index\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    627\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    628\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice_memory_limit\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cuda/utils.py\u001b[0m in \u001b[0;36mget_device_total_memory\u001b[0;34m(index)\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misnumeric\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    235\u001b[0m         \u001b[0;31m# This means index is UUID. This works for both MIG and non-MIG device UUIDs.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 236\u001b[0;31m         \u001b[0mhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpynvml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnvmlDeviceGetHandleByUUID\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    237\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    238\u001b[0m         \u001b[0;31m# This is a device index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/pynvml/nvml.py\u001b[0m in \u001b[0;36mnvmlDeviceGetHandleByUUID\u001b[0;34m(uuid)\u001b[0m\n\u001b[1;32m   1669\u001b[0m     \u001b[0mfn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_nvmlGetFunctionPointer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"nvmlDeviceGetHandleByUUID\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1670\u001b[0m     \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc_uuid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbyref\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1671\u001b[0;31m     \u001b[0m_nvmlCheckReturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1672\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdevice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1673\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/pynvml/nvml.py\u001b[0m in \u001b[0;36m_nvmlCheckReturn\u001b[0;34m(ret)\u001b[0m\n\u001b[1;32m    763\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_nvmlCheckReturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    764\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mNVML_SUCCESS\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 765\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mNVMLError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mret\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    766\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mret\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    767\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNVMLError_NotFound\u001b[0m: Not Found"
     ]
    }
   ],
   "source": [
    "# This will use all GPUs on the local host by default\n",
    "cluster = LocalCUDACluster(threads_per_worker=1)\n",
    "c = Client(cluster)\n",
    "\n",
    "# Query the client for all connected workers\n",
    "workers = c.has_what().keys()\n",
    "n_workers = len(workers)\n",
    "n_streams = 8 # Performance optimization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define Parameters\n",
    "\n",
    "In addition to the number of examples, random forest fitting performance depends heavily on the number of columns in a dataset and (especially) on the maximum depth to which trees are allowed to grow. Lower `max_depth` values can greatly speed up fitting, though going too low may reduce accuracy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random Forest building parameters\n",
    "max_depth = 12\n",
    "n_bins = 16\n",
    "n_trees = 1000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribute data to worker GPUs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = X_train.astype(np.float32)\n",
    "X_test = X_test.astype(np.float32)\n",
    "y_train = y_train.astype(np.int32)\n",
    "y_test = y_test.astype(np.int32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_partitions = n_workers\n",
    "\n",
    "def distribute(X, y):\n",
    "    # First convert to cudf (with real data, you would likely load in cuDF format to start)\n",
    "    X_cudf = cudf.DataFrame.from_pandas(pd.DataFrame(X))\n",
    "    y_cudf = cudf.Series(y)\n",
    "\n",
    "    # Partition with Dask\n",
    "    # In this case, each worker will train on 1/n_partitions fraction of the data\n",
    "    X_dask = dask_cudf.from_cudf(X_cudf, npartitions=n_partitions)\n",
    "    y_dask = dask_cudf.from_cudf(y_cudf, npartitions=n_partitions)\n",
    "\n",
    "    # Persist to cache the data in active memory\n",
    "    X_dask, y_dask = \\\n",
    "      dask_utils.persist_across_workers(c, [X_dask, y_dask], workers=workers)\n",
    "    \n",
    "    return X_dask, y_dask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Modify the code in this cell\n",
    "\n",
    "X_train_dask, y_train_dask = distribute(X_train, y_train) #Pass train data as arguments here\n",
    "X_test_dask, y_test_dask = distribute(X_test, y_test) #Pass test data as arguments here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create the  Scikit-learn model\n",
    "\n",
    "Since a scikit-learn equivalent to the multi-node multi-GPU K-means in cuML doesn't exist, we will use Dask-ML's implementation for comparison."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 206 ms, sys: 16.8 ms, total: 223 ms\n",
      "Wall time: 242 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RandomForestClassifier(max_depth=12, n_estimators=150, n_jobs=-1)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "max_depth = 12\n",
    "n_trees = 150\n",
    "\n",
    "# Use all avilable CPU cores\n",
    "skl_model = sklRF(n_estimators=n_trees, max_depth=max_depth, n_jobs=-1)\n",
    "skl_model.fit(X_train, y_train.iloc[:,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xgboost as xgb\n",
    "xgr=xgb.XGBClassifier(max_depth=50,min_child_weight=0.1,gamma=0.0001)\n",
    "dtrain = xgb.DMatrix(X_train, label=y_train.iloc[:,1])\n",
    "dtest = xgb.DMatrix(X_test, label=y_test.iloc[:,1])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tree_method': 'gpu_hist', 'objective': 'reg:squarederror'}\n"
     ]
    }
   ],
   "source": [
    "# instantiate params\n",
    "params = {}\n",
    "\n",
    "# general params\n",
    "general_params = {}\n",
    "params.update(general_params)\n",
    "\n",
    "# booster params\n",
    "booster_params = {'tree_method': 'gpu_hist'}\n",
    "params.update(booster_params)\n",
    "\n",
    "# learning task params\n",
    "learning_task_params = {'objective': 'reg:squarederror'}\n",
    "params.update(learning_task_params)\n",
    "print(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "evallist = [(dtest, 'test'), (dtrain, 'train')]\n",
    "num_round = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0]\ttest-rmse:0.39144\ttrain-rmse:0.35744\n",
      "[1]\ttest-rmse:0.32951\ttrain-rmse:0.25557\n",
      "[2]\ttest-rmse:0.29880\ttrain-rmse:0.18276\n",
      "[3]\ttest-rmse:0.28634\ttrain-rmse:0.13072\n",
      "[4]\ttest-rmse:0.28309\ttrain-rmse:0.09351\n",
      "[5]\ttest-rmse:0.28381\ttrain-rmse:0.06691\n",
      "[6]\ttest-rmse:0.28591\ttrain-rmse:0.04788\n",
      "[7]\ttest-rmse:0.28821\ttrain-rmse:0.03427\n",
      "[8]\ttest-rmse:0.29027\ttrain-rmse:0.02453\n",
      "[9]\ttest-rmse:0.29195\ttrain-rmse:0.01757\n",
      "[10]\ttest-rmse:0.29327\ttrain-rmse:0.01258\n",
      "[11]\ttest-rmse:0.29427\ttrain-rmse:0.00901\n",
      "[12]\ttest-rmse:0.29502\ttrain-rmse:0.00645\n",
      "[13]\ttest-rmse:0.29557\ttrain-rmse:0.00462\n",
      "[14]\ttest-rmse:0.29597\ttrain-rmse:0.00331\n",
      "[15]\ttest-rmse:0.29627\ttrain-rmse:0.00238\n",
      "[16]\ttest-rmse:0.29648\ttrain-rmse:0.00170\n",
      "[17]\ttest-rmse:0.29664\ttrain-rmse:0.00122\n",
      "[18]\ttest-rmse:0.29675\ttrain-rmse:0.00088\n",
      "[19]\ttest-rmse:0.29683\ttrain-rmse:0.00063\n",
      "[20]\ttest-rmse:0.29689\ttrain-rmse:0.00045\n",
      "[21]\ttest-rmse:0.29693\ttrain-rmse:0.00032\n",
      "[22]\ttest-rmse:0.29696\ttrain-rmse:0.00023\n",
      "[23]\ttest-rmse:0.29699\ttrain-rmse:0.00017\n",
      "[24]\ttest-rmse:0.29699\ttrain-rmse:0.00016\n",
      "[25]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[26]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[27]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[28]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[29]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[30]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[31]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[32]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[33]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[34]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[35]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[36]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[37]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[38]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[39]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[40]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[41]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[42]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[43]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[44]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[45]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[46]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[47]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[48]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[49]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[50]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[51]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[52]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[53]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[54]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[55]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[56]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[57]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[58]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[59]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[60]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[61]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[62]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[63]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[64]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[65]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[66]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[67]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[68]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[69]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[70]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[71]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[72]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[73]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[74]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[75]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[76]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[77]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[78]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[79]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[80]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[81]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[82]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[83]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[84]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[85]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[86]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[87]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[88]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[89]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[90]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[91]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[92]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[93]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[94]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[95]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[96]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[97]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[98]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[99]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[100]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[101]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[102]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[103]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[104]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[105]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[106]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[107]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[108]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[109]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[110]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[111]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[112]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[113]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[114]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[115]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[116]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[117]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[118]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[119]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[120]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[121]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[122]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[123]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[124]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[125]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[126]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[127]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[128]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[129]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[130]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[131]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[132]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[133]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[134]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[135]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[136]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[137]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[138]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[139]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[140]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[141]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[142]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[143]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[144]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[145]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[146]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[147]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[148]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[149]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[150]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[151]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[152]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[153]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[154]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[155]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[156]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[157]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[158]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[159]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[160]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[161]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[162]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[163]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[164]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[165]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[166]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[167]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[168]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[169]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[170]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[171]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[172]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[173]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[174]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[175]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[176]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[177]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[178]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[179]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[180]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[181]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[182]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[183]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[184]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[185]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[186]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[187]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[188]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[189]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[190]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[191]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[192]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[193]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[194]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[195]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[196]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[197]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[198]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[199]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[200]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[201]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[202]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[203]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[204]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[205]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[206]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[207]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[208]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[209]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[210]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[211]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[212]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[213]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[214]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[215]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[216]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[217]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[218]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[219]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[220]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[221]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[222]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[223]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[224]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[225]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[226]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[227]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[228]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[229]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[230]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[231]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[232]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[233]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[234]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[235]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[236]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[237]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[238]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[239]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[240]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[241]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[242]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[243]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[244]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[245]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[246]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[247]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[248]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[249]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[250]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[251]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[252]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[253]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[254]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[255]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[256]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[257]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[258]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[259]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[260]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[261]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[262]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[263]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[264]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[265]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[266]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[267]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[268]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[269]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[270]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[271]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[272]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[273]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[274]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[275]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[276]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[277]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[278]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[279]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[280]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[281]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[282]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[283]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[284]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[285]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[286]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[287]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[288]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[289]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[290]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[291]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[292]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[293]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[294]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[295]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[296]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[297]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[298]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n",
      "[299]\ttest-rmse:0.29698\ttrain-rmse:0.00016\n"
     ]
    }
   ],
   "source": [
    "bst = xgb.train(params, dtrain, num_round, evallist)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred_train = bst.predict(dtrain)\n",
    "y_pred_train = np.round(y_pred_train).astype(np.int32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1,\n",
       "       1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0], dtype=int32)"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XGboost accuracy:   1.0\n"
     ]
    }
   ],
   "source": [
    "print(\"XGboost accuracy:  \", accuracy_score(y_train.iloc[:,1], y_pred_train))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XGboost accuracy:   0.9117647058823529\n"
     ]
    }
   ],
   "source": [
    "y_pred_test = bst.predict(dtest)\n",
    "y_pred_test = np.round(y_pred_test).astype(np.int32)\n",
    "print(\"XGboost accuracy:  \", accuracy_score(y_test.iloc[:,1], y_pred_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     0\n",
       "1     0\n",
       "2     0\n",
       "3     0\n",
       "4     0\n",
       "5     0\n",
       "6     0\n",
       "7     0\n",
       "8     0\n",
       "9     0\n",
       "10    0\n",
       "11    1\n",
       "12    1\n",
       "13    1\n",
       "14    1\n",
       "15    1\n",
       "16    0\n",
       "17    0\n",
       "18    1\n",
       "19    1\n",
       "20    0\n",
       "21    1\n",
       "22    1\n",
       "23    1\n",
       "24    1\n",
       "25    1\n",
       "26    1\n",
       "27    1\n",
       "28    0\n",
       "29    0\n",
       "30    0\n",
       "31    0\n",
       "32    0\n",
       "33    0\n",
       "Name: cancer, dtype: int32"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_test.iloc[:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-71-8249994f44f4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0my_test_cp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfloat32\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"SKLearn accuracy:  \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccuracy_score\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred_cp\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0mextra_args\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mextra_args\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m             \u001b[0;31m# extra_args > 0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36maccuracy_score\u001b[0;34m(y_true, y_pred, normalize, sample_weight)\u001b[0m\n\u001b[1;32m    200\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m     \u001b[0;31m# Compute accuracy for each possible representation\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 202\u001b[0;31m     \u001b[0my_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_check_targets\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    203\u001b[0m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_weight\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    204\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0my_type\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'multilabel'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36m_check_targets\u001b[0;34m(y_true, y_pred)\u001b[0m\n\u001b[1;32m     83\u001b[0m     \u001b[0mcheck_consistent_length\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     84\u001b[0m     \u001b[0mtype_true\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 85\u001b[0;31m     \u001b[0mtype_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtype_of_target\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     86\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m     \u001b[0my_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mtype_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype_pred\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mtype_of_target\u001b[0;34m(y)\u001b[0m\n\u001b[1;32m    259\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"y cannot be class 'SparseSeries' or 'SparseArray'\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 261\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mis_multilabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    262\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m'multilabel-indicator'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/sklearn/utils/multiclass.py\u001b[0m in \u001b[0;36mis_multilabel\u001b[0;34m(y)\u001b[0m\n\u001b[1;32m    145\u001b[0m             \u001b[0mwarnings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msimplefilter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'error'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mVisibleDeprecationWarning\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    146\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 147\u001b[0;31m                 \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    148\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mVisibleDeprecationWarning\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m                 \u001b[0;31m# dtype=object should be provided explicitly for ragged arrays,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/envs/rapids/lib/python3.7/site-packages/numpy/core/_asarray.py\u001b[0m in \u001b[0;36masarray\u001b[0;34m(a, dtype, order, like)\u001b[0m\n\u001b[1;32m    100\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0m_asarray_with_like\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlike\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlike\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 102\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morder\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    103\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    104\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mcupy/_core/core.pyx\u001b[0m in \u001b[0;36mcupy._core.core.ndarray.__array__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: Implicit conversion to a NumPy array is not allowed. Please use `.get()` to construct a NumPy array explicitly."
     ]
    }
   ],
   "source": [
    "import cupy as cp\n",
    "from cuml.metrics.regression import r2_score\n",
    "\n",
    "y_pred_test = bst.predict(dtest)\n",
    "\n",
    "y_pred_cp = cp.asarray(y_pred_test)\n",
    "y_test_cp = cp.asarray(y_train).astype(np.float32)\n",
    "\n",
    "print(\"SKLearn accuracy:  \", accuracy_score(y_test.iloc[:,1], y_pred_cp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GridSearchCV(estimator=RandomForestClassifier(),\n",
       "             param_grid={'max_depth': [10, 15, 50, 100],\n",
       "                         'n_estimators': [500, 1000, 1500, 2000]},\n",
       "             scoring='accuracy')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.model_selection import GridSearchCV\n",
    "params_grid = {'max_depth': [10, 15, 50, 100], 'n_estimators': [500, 1000, 1500, 2000]}\n",
    "grid = GridSearchCV(sklRF(), params_grid, scoring='accuracy')\n",
    "grid.fit(X_train, y_train.iloc[:,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'max_depth': 10, 'n_estimators': 500}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grid.best_params_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Train the distributed cuML model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Line magic function `%%time` not found.\n"
     ]
    }
   ],
   "source": [
    "#Modify the code in this cell\n",
    "\n",
    "%%time\n",
    "\n",
    "cuml_model = cumlDaskRF(max_depth=max_depth, n_estimators=n_trees, n_bins=n_bins, n_streams=n_streams)\n",
    "cuml_model.fit() # Pass X and y train dask data here\n",
    "\n",
    "wait(cuml_model.rfs) # Allow asynchronous training tasks to finish"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Predict and check accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SKLearn accuracy:   1.0\n"
     ]
    }
   ],
   "source": [
    "#Modify the code in this cell\n",
    "\n",
    "skl_y_pred = skl_model.predict(X_train)\n",
    "#cuml_y_pred = cuml_model.predict().compute().to_array()  #Pass the X test dask data as argument here\n",
    "\n",
    "# Due to randomness in the algorithm, you may see slight variation in accuracies\n",
    "print(\"SKLearn accuracy:  \", accuracy_score(y_train.iloc[:,1], skl_y_pred))\n",
    "#print(\"CuML accuracy:     \", accuracy_score())  #Pass the y test dask data  and predicted values from CuML model as argument here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "skl_y_pred"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='ex4'></a><br>\n",
    "\n",
    "### 4. CONCLUSION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare the performance of our solution!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Algorithm     | Implementation | Accuracy      | Time | Algorithm     | Implementation | Accuracy      | Time |\n",
    "| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write down your observations and compare the CuML and Scikit learn scores. They should be approximately equal.  We hope that you found this exercise exciting and beneficial in understanding RAPIDS better. Share your highest accuracy and try to use the unique features of RAPIDS for accelerating your data science pipelines. Don't restrict yourself to the previously explained concepts, but use the documentation to apply more models and functions and achieve the best results. Jump over to the next notebook for our sample solution.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5.  References\n",
    "\n",
    "\n",
    "\n",
    "<p xmlns:dct=\"http://purl.org/dc/terms/\">\n",
    "  <a rel=\"license\"\n",
    "     href=\"http://creativecommons.org/publicdomain/zero/1.0/\">\n",
    "    <center><img src=\"http://i.creativecommons.org/p/zero/1.0/88x31.png\" style=\"border-style: none;\" alt=\"CC0\"  /></center>\n",
    "  </a>\n",
    " \n",
    "</p>\n",
    "\n",
    "\n",
    "- The dataset is licensed under a CC0: Public Domain license.\n",
    "\n",
    "- Molecular Classification of Cancer: Class Discovery and Class Prediction by Gene Expression. Science 286:531-537. (1999). Published: 1999.10.14. T.R. Golub, D.K. Slonim, P. Tamayo, C. Huard, M. Gaasenbeek, J.P. Mesirov, H. Coller, M. Loh, J.R. Downing, M.A. Caligiuri, C.D. Bloomfield, and E.S. Lander\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Licensing\n",
    "  \n",
    "This material is released by OpenACC-Standard.org, in collaboration with NVIDIA Corporation, under the Creative Commons Attribution 4.0 International (CC BY 4.0)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;&emsp;&emsp;\n",
    "&emsp;&emsp;&emsp;\n",
    "[Home Page](../../START_HERE.ipynb)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
